# Data Description

- read_name - "The original read name"
- read_length - "The length of the read in bases"
- read_idx - "Unique integer ID of the read"
- contact_is_direct - "There are no intervening assigned restriction fragments on the read"
- contact_is_cis - "Both alignments come from the same chromsome/contig"
- contact_read_distance - "The distance between the end of the left alignment and the start of the right alignment on the read"
- contact_genome_distance - "The distance between the end of the left alignment and the start of the right alignment (valid for cis contacts only)"
- contact_fragment_adjacent - "A boolean to indicate if the contact is between the same or adjacent fragments"
- contact_fragment_distance - "The distance between the midpoints of the assigned fragments (valid for cis contacts only)"
- haplotype_pair_type - "A categorical variable describing the relationship between the haplotypes assigned to each of the alignments in a contact"
- align1_align_idx - "Unique integer ID of the first aligned segment"
- align1_chrom - "The chromosome/contig of the first aligned segment"
- align1_start - "The zero-based start position on the genome of the alignment"
- align1_end - "The end position on the genome of the alignment"
- align1_strand - "The alignment strand"
- align1_mapping_quality - "The mapping quality as calculated by the aligner"
- align1_align_score - "The alignment score as calculated by the aligner"
- align1_align_base_qscore - "The mean read base score for the aligned segment (rounded to the nearest integer)."
- align1_phase_set - "The ID of the phase set, often this is the start position of the phase block"
- align1_phase_qual - "The phred-scaled quality score of the haplotype assignment"
- align1_haplotype - "The id of the haplotype within this block, usually set to 1 or 2. A value of -1 means that this alignment is unphased"
- align1_fragment_id - "The UID of the restriction fragment assigned to this alignment"
- align1_fragment_start - "The start point on the genome of this restriction fragment"
- align1_fragment_end - "The end point on the genome of this restriction fragment"
- align2_align_idx - "Unique integer ID of the first aligned segment"
- align2_chrom - "The chromosome/contig of the first aligned segment"
- align2_start - "The zero-based start position on the genome of the alignment"
- align2_end - "The end position on the genome of the alignment"
- align2_strand - "The alignment strand"
- align2_mapping_quality - The mapping quality as calculated by the aligner"
- align2_align_score - "The alignment score as calculated by the aligner",
- align2_align_base_qscore= - "The mean read base score for the aligned segment (rounded to the nearest integer)."
- align2_phase_set - "The ID of the phase set, often this is the start position of the phase block"
- align2_phase_qual - "The phred-scaled quality score of the haplotype assignment"
- align2_haplotype - "The id of the haplotype within this block, usually set to 1 or 2. A value of -1 means that this alignment is unphased"
- align2_fragment_id - "The UID of the restriction fragment assigned to this alignment"
- align2_fragment_start - "The start point on the genome of this restriction fragment"
- align2_fragment_end - "The end point on the genome of this restriction fragment"