{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import cooler\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors\n",
    "from matplotlib.patches import Patch\n",
    "from matplotlib.offsetbox import AnchoredText\n",
    "import scipy.io\n",
    "import pyBigWig\n",
    "import gget\n",
    "import pysam\n",
    "\n",
    "import os\n",
    "import sys\n",
    "from importlib import reload\n",
    "\n",
    "import filters\n",
    "import pore_c_utils as pcu\n",
    "import binning_utils as binning\n",
    "import plotting as hicPlot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "assembly = pcu.loadAssembly(\"GRCm39_Assembly.txt\")\n",
    "\n",
    "chromDict = dict(zip(assembly['RefSeq accession'],assembly['Chromosome'].apply(lambda x : f\"chr{x}\")))\n",
    "\n",
    "# the reverse for easy lookup\n",
    "chromDict_r = {v: k for k, v in chromDict.items()}\n",
    "chromOrder = list(chromDict.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Filename</th>\n",
       "      <th>Experiment Accession</th>\n",
       "      <th>File Accession</th>\n",
       "      <th>Size (MB)</th>\n",
       "      <th>File Type</th>\n",
       "      <th>File Format</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4DNFI7FO9H19.bam</td>\n",
       "      <td>4DNEXCHOKPKJ</td>\n",
       "      <td>4DNFI7FO9H19</td>\n",
       "      <td>1503.75</td>\n",
       "      <td>read positions</td>\n",
       "      <td>bam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4DNFI8CSCJWM.tsv</td>\n",
       "      <td>4DNEXCHOKPKJ</td>\n",
       "      <td>4DNFI8CSCJWM</td>\n",
       "      <td>23.19</td>\n",
       "      <td>isoform expression</td>\n",
       "      <td>tsv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4DNFIUW8CG2I.bw</td>\n",
       "      <td>4DNEXCHOKPKJ</td>\n",
       "      <td>4DNFIUW8CG2I</td>\n",
       "      <td>42.86</td>\n",
       "      <td>read counts (plus)</td>\n",
       "      <td>bw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4DNFIXOTRTRM.bw</td>\n",
       "      <td>4DNEXCHOKPKJ</td>\n",
       "      <td>4DNFIXOTRTRM</td>\n",
       "      <td>42.06</td>\n",
       "      <td>read counts (minus)</td>\n",
       "      <td>bw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4DNFIYTCHMIZ.tsv</td>\n",
       "      <td>4DNEXCHOKPKJ</td>\n",
       "      <td>4DNFIYTCHMIZ</td>\n",
       "      <td>11.23</td>\n",
       "      <td>gene expression</td>\n",
       "      <td>tsv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4DNFI12AUKQS.bw</td>\n",
       "      <td>4DNEXGMSIOHL</td>\n",
       "      <td>4DNFI12AUKQS</td>\n",
       "      <td>31.51</td>\n",
       "      <td>read counts (plus)</td>\n",
       "      <td>bw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4DNFI3YYNDKI.tsv</td>\n",
       "      <td>4DNEXGMSIOHL</td>\n",
       "      <td>4DNFI3YYNDKI</td>\n",
       "      <td>11.23</td>\n",
       "      <td>gene expression</td>\n",
       "      <td>tsv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4DNFIFVPB94O.bw</td>\n",
       "      <td>4DNEXGMSIOHL</td>\n",
       "      <td>4DNFIFVPB94O</td>\n",
       "      <td>30.93</td>\n",
       "      <td>read counts (minus)</td>\n",
       "      <td>bw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>4DNFIPYGE7JR.tsv</td>\n",
       "      <td>4DNEXGMSIOHL</td>\n",
       "      <td>4DNFIPYGE7JR</td>\n",
       "      <td>23.18</td>\n",
       "      <td>isoform expression</td>\n",
       "      <td>tsv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>4DNFISUXPL3L.bam</td>\n",
       "      <td>4DNEXGMSIOHL</td>\n",
       "      <td>4DNFISUXPL3L</td>\n",
       "      <td>1063.16</td>\n",
       "      <td>read positions</td>\n",
       "      <td>bam</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Filename Experiment Accession File Accession  Size (MB)  \\\n",
       "0  4DNFI7FO9H19.bam         4DNEXCHOKPKJ   4DNFI7FO9H19    1503.75   \n",
       "1  4DNFI8CSCJWM.tsv         4DNEXCHOKPKJ   4DNFI8CSCJWM      23.19   \n",
       "2   4DNFIUW8CG2I.bw         4DNEXCHOKPKJ   4DNFIUW8CG2I      42.86   \n",
       "3   4DNFIXOTRTRM.bw         4DNEXCHOKPKJ   4DNFIXOTRTRM      42.06   \n",
       "4  4DNFIYTCHMIZ.tsv         4DNEXCHOKPKJ   4DNFIYTCHMIZ      11.23   \n",
       "5   4DNFI12AUKQS.bw         4DNEXGMSIOHL   4DNFI12AUKQS      31.51   \n",
       "6  4DNFI3YYNDKI.tsv         4DNEXGMSIOHL   4DNFI3YYNDKI      11.23   \n",
       "7   4DNFIFVPB94O.bw         4DNEXGMSIOHL   4DNFIFVPB94O      30.93   \n",
       "8  4DNFIPYGE7JR.tsv         4DNEXGMSIOHL   4DNFIPYGE7JR      23.18   \n",
       "9  4DNFISUXPL3L.bam         4DNEXGMSIOHL   4DNFISUXPL3L    1063.16   \n",
       "\n",
       "             File Type File Format  \n",
       "0       read positions         bam  \n",
       "1   isoform expression         tsv  \n",
       "2   read counts (plus)          bw  \n",
       "3  read counts (minus)          bw  \n",
       "4      gene expression         tsv  \n",
       "5   read counts (plus)          bw  \n",
       "6      gene expression         tsv  \n",
       "7  read counts (minus)          bw  \n",
       "8   isoform expression         tsv  \n",
       "9       read positions         bam  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dirpath = \"/nfs/turbo/umms-indikar/shared/projects/poreC/nagano_2017_rnaseq/\"\n",
    "\n",
    "filepath = f\"{dirpath}4DNESZ741PJM_processed_files_2022-05-30-14h-43m.tsv\"\n",
    "summary = pd.read_csv(filepath, sep='\\t', comment='#')\n",
    "\n",
    "summary['Filename'] = summary['File Download URL'].apply(lambda x: x.split(\"/\")[-1])\n",
    "cols = ['Filename', 'Experiment Accession', 'File Accession', 'Size (MB)', 'File Type', 'File Format',]\n",
    "summary[cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id(s)</th>\n",
       "      <th>length</th>\n",
       "      <th>effective_length</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>posterior_mean_count</th>\n",
       "      <th>posterior_standard_deviation_of_count</th>\n",
       "      <th>pme_TPM</th>\n",
       "      <th>pme_FPKM</th>\n",
       "      <th>TPM_ci_lower_bound</th>\n",
       "      <th>TPM_ci_upper_bound</th>\n",
       "      <th>TPM_coefficient_of_quartile_variation</th>\n",
       "      <th>FPKM_ci_lower_bound</th>\n",
       "      <th>FPKM_ci_upper_bound</th>\n",
       "      <th>FPKM_coefficient_of_quartile_variation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10000</td>\n",
       "      <td>10000</td>\n",
       "      <td>72.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10001</td>\n",
       "      <td>10001</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10002</td>\n",
       "      <td>10002</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10003</td>\n",
       "      <td>10003</td>\n",
       "      <td>75.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10004</td>\n",
       "      <td>10004</td>\n",
       "      <td>78.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_id transcript_id(s)  length  effective_length  expected_count  TPM  \\\n",
       "0   10000            10000    72.0               0.0             0.0  0.0   \n",
       "1   10001            10001    73.0               0.0             0.0  0.0   \n",
       "2   10002            10002    73.0               0.0             0.0  0.0   \n",
       "3   10003            10003    75.0               0.0             0.0  0.0   \n",
       "4   10004            10004    78.0               0.0             0.0  0.0   \n",
       "\n",
       "   FPKM  posterior_mean_count  posterior_standard_deviation_of_count  pme_TPM  \\\n",
       "0   0.0                   0.0                                    0.0      0.0   \n",
       "1   0.0                   0.0                                    0.0      0.0   \n",
       "2   0.0                   0.0                                    0.0      0.0   \n",
       "3   0.0                   0.0                                    0.0      0.0   \n",
       "4   0.0                   0.0                                    0.0      0.0   \n",
       "\n",
       "   pme_FPKM  TPM_ci_lower_bound  TPM_ci_upper_bound  \\\n",
       "0       0.0                 0.0                 0.0   \n",
       "1       0.0                 0.0                 0.0   \n",
       "2       0.0                 0.0                 0.0   \n",
       "3       0.0                 0.0                 0.0   \n",
       "4       0.0                 0.0                 0.0   \n",
       "\n",
       "   TPM_coefficient_of_quartile_variation  FPKM_ci_lower_bound  \\\n",
       "0                                    0.0                  0.0   \n",
       "1                                    0.0                  0.0   \n",
       "2                                    0.0                  0.0   \n",
       "3                                    0.0                  0.0   \n",
       "4                                    0.0                  0.0   \n",
       "\n",
       "   FPKM_ci_upper_bound  FPKM_coefficient_of_quartile_variation  \n",
       "0                  0.0                                     0.0  \n",
       "1                  0.0                                     0.0  \n",
       "2                  0.0                                     0.0  \n",
       "3                  0.0                                     0.0  \n",
       "4                  0.0                                     0.0  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filepath = f\"{dirpath}4DNFIYTCHMIZ.tsv\"\n",
    "\n",
    "df = pd.read_csv(filepath, sep='\\t')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df.shape=(81881, 17)\n",
      "df.shape=(19040, 17)\n"
     ]
    }
   ],
   "source": [
    "print(f\"{df.shape=}\")\n",
    "\n",
    "df = df[df['TPM'] > 0]\n",
    "# df = df[df['gene_id'].str.contains(\"ENSMUSG\")]\n",
    "print(f\"{df.shape=}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31383                    1\n",
       "ENSMUSG00000063410.8     1\n",
       "ENSMUSG00000063383.5     1\n",
       "ENSMUSG00000063382.6     1\n",
       "ENSMUSG00000063364.10    1\n",
       "                        ..\n",
       "ENSMUSG00000030770.15    1\n",
       "ENSMUSG00000030769.15    1\n",
       "ENSMUSG00000030768.12    1\n",
       "ENSMUSG00000030766.15    1\n",
       "ENSMUSG00000118386.1     1\n",
       "Name: gene_id, Length: 19040, dtype: int64"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['gene_id'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # fetch data\n",
    "# geneInfo = gget.info(df['gene_id'], expand=True)\n",
    "\n",
    "# geneInfo.to_csv(\"filtered_data/geneNames.csv\", index=False)\n",
    "# print('done.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
