{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from collections import Counter\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import math\n",
    "from scipy import stats\n",
    "from matplotlib import transforms\n",
    "import networkx as nx\n",
    "from matplotlib.gridspec import GridSpec\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "from scipy.sparse import csgraph\n",
    "\n",
    "\n",
    "import os\n",
    "import sys\n",
    "from importlib import reload\n",
    "\n",
    "# local imports \n",
    "import filters\n",
    "import pore_c_utils as pcu\n",
    "import binning_utils as binning\n",
    "import plotting as hicPlot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'filters' from '/home/cstansbu/git_repositories/scPoreC-analysis/notebooks/filters.py'>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reload(pcu)\n",
    "reload(binning)\n",
    "reload(filters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "runId='run01' df.shape=(13386738, 38)\n",
      "runId='run02' df.shape=(13046898, 38)\n",
      "runId='run04' df.shape=(20486346, 38)\n",
      "runId='run05' df.shape=(5043403, 38)\n",
      "runId='run06' df.shape=(7310826, 38)\n",
      "runId='run07' df.shape=(10525887, 38)\n",
      "rawCells.keys()=dict_keys(['run01', 'run02', 'run04', 'run05', 'run06', 'run07'])\n"
     ]
    }
   ],
   "source": [
    "dirpath = \"/nfs/turbo/umms-indikar/shared/tools/Pore-C-Snakemake/results_ES5to12/contacts/\"\n",
    "cellNums = [1, 2, 4, 5, 6, 7] # dropping cell 3 due to sparsity\n",
    "filetags = [f\"run0{x}\" for x in cellNums]\n",
    "\n",
    "# filetags = [f\"run0{x}\" for x in range(1, 2)]\n",
    "\n",
    "assembly = pcu.loadAssembly(\"GRCm39_Assembly.txt\")\n",
    "rawCells = pcu.loadRawContactDirectory(dirpath, filetags)\n",
    "print(f\"{rawCells.keys()=}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------------------\n",
      "runId='run01'\n",
      "cf.shape=(5406, 64)\n",
      "-- -- --\n",
      "Total Reads 3369 (1.000)\n",
      "Singletons 2236 (0.664)\n",
      "Pairs 669 (0.199)\n",
      "Multiway 464 (0.138)\n",
      "-- -- --\n",
      "Total Contacts 5406 (1.000)\n",
      "cis Contacts 5406 (1.000)\n",
      "trans Contacts 0 (0.000)\n",
      "-- -- --\n",
      "Total Contacts 5406 (1.000)\n",
      "Direct Contacts 5406 (1.000)\n",
      "Indirect Contacts 0 (0.000)\n",
      "\n",
      "------------------------------\n",
      "runId='run02'\n",
      "cf.shape=(2689, 64)\n",
      "-- -- --\n",
      "Total Reads 1668 (1.000)\n",
      "Singletons 1091 (0.654)\n",
      "Pairs 350 (0.210)\n",
      "Multiway 227 (0.136)\n",
      "-- -- --\n",
      "Total Contacts 2689 (1.000)\n",
      "cis Contacts 2689 (1.000)\n",
      "trans Contacts 0 (0.000)\n",
      "-- -- --\n",
      "Total Contacts 2689 (1.000)\n",
      "Direct Contacts 2689 (1.000)\n",
      "Indirect Contacts 0 (0.000)\n",
      "\n",
      "------------------------------\n",
      "runId='run04'\n",
      "cf.shape=(2735, 64)\n",
      "-- -- --\n",
      "Total Reads 1693 (1.000)\n",
      "Singletons 1103 (0.652)\n",
      "Pairs 358 (0.211)\n",
      "Multiway 232 (0.137)\n",
      "-- -- --\n",
      "Total Contacts 2735 (1.000)\n",
      "cis Contacts 2735 (1.000)\n",
      "trans Contacts 0 (0.000)\n",
      "-- -- --\n",
      "Total Contacts 2735 (1.000)\n",
      "Direct Contacts 2735 (1.000)\n",
      "Indirect Contacts 0 (0.000)\n",
      "\n",
      "------------------------------\n",
      "runId='run05'\n",
      "cf.shape=(3774, 64)\n",
      "-- -- --\n",
      "Total Reads 2211 (1.000)\n",
      "Singletons 1316 (0.595)\n",
      "Pairs 551 (0.249)\n",
      "Multiway 344 (0.156)\n",
      "-- -- --\n",
      "Total Contacts 3774 (1.000)\n",
      "cis Contacts 3774 (1.000)\n",
      "trans Contacts 0 (0.000)\n",
      "-- -- --\n",
      "Total Contacts 3774 (1.000)\n",
      "Direct Contacts 3774 (1.000)\n",
      "Indirect Contacts 0 (0.000)\n",
      "\n",
      "------------------------------\n",
      "runId='run06'\n",
      "cf.shape=(3705, 64)\n",
      "-- -- --\n",
      "Total Reads 2245 (1.000)\n",
      "Singletons 1396 (0.622)\n",
      "Pairs 514 (0.229)\n",
      "Multiway 335 (0.149)\n",
      "-- -- --\n",
      "Total Contacts 3705 (1.000)\n",
      "cis Contacts 3705 (1.000)\n",
      "trans Contacts 0 (0.000)\n",
      "-- -- --\n",
      "Total Contacts 3705 (1.000)\n",
      "Direct Contacts 3705 (1.000)\n",
      "Indirect Contacts 0 (0.000)\n",
      "\n",
      "------------------------------\n",
      "runId='run07'\n",
      "cf.shape=(5513, 64)\n",
      "-- -- --\n",
      "Total Reads 3435 (1.000)\n",
      "Singletons 2226 (0.648)\n",
      "Pairs 732 (0.213)\n",
      "Multiway 477 (0.139)\n",
      "-- -- --\n",
      "Total Contacts 5513 (1.000)\n",
      "cis Contacts 5513 (1.000)\n",
      "trans Contacts 0 (0.000)\n",
      "-- -- --\n",
      "Total Contacts 5513 (1.000)\n",
      "Direct Contacts 5513 (1.000)\n",
      "Indirect Contacts 0 (0.000)\n",
      "\n",
      "filteredCells.keys()=dict_keys(['run01', 'run02', 'run04', 'run05', 'run06', 'run07'])\n"
     ]
    }
   ],
   "source": [
    "reload(pcu)\n",
    "reload(binning)\n",
    "reload(filters)\n",
    "\n",
    "filteredCells = filters.runAllFilters(rawCells,\n",
    "                                      assembly,\n",
    "                                      adjacent=True,\n",
    "                                      chromosome='NC_000068.8', \n",
    "                                      verbose=True)  \n",
    "\n",
    "print(f\"{filteredCells.keys()=}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# make interaction networks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "runId='run01'\n",
      "A.shape=(182, 182)\n",
      "np.max(A)=2.0\n",
      "\n",
      "runId='run02'\n",
      "A.shape=(182, 182)\n",
      "np.max(A)=2.0\n",
      "\n",
      "runId='run04'\n",
      "A.shape=(182, 182)\n",
      "np.max(A)=2.0\n",
      "\n",
      "runId='run05'\n",
      "A.shape=(182, 182)\n",
      "np.max(A)=2.0\n",
      "\n",
      "runId='run06'\n",
      "A.shape=(182, 182)\n",
      "np.max(A)=2.0\n",
      "\n",
      "runId='run07'\n",
      "A.shape=(182, 182)\n",
      "np.max(A)=2.0\n",
      "\n",
      "dict_keys(['run01', 'run02', 'run04', 'run05', 'run06', 'run07'])\n"
     ]
    }
   ],
   "source": [
    "reload(pcu)\n",
    "reload(binning)\n",
    "\n",
    "chomosome = 2\n",
    "chromLength = assembly.loc[assembly['Chromosome'] == str(chomosome), 'Total length'].values[0]\n",
    "\n",
    "resolution = 1000000 # 1mb scale\n",
    "chromBins = binning.getBins(chromLength, resolution)\n",
    "\n",
    "label1 = 'bin1'\n",
    "label2 = 'bin2'\n",
    "\n",
    "matrices = {}\n",
    "\n",
    "for runId, data in filteredCells.items():\n",
    "    print(f\"{runId=}\")\n",
    "    \n",
    "    chrom = binning.chromosomeBinData(data, \n",
    "                                     chromBins, \n",
    "                                     label1=label1, \n",
    "                                     label2=label2)\n",
    "    \n",
    "    A = pcu.getHic(chrom, chromBins, label1, label2)\n",
    "    print(f\"{A.shape=}\")\n",
    "    print(f\"{np.max(A)=}\")\n",
    "    matrices[runId] = A\n",
    "    \n",
    "    print()\n",
    "    \n",
    "print(f\"{matrices.keys()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 5, 7\n",
    "\n",
    "\n",
    "def sortedEigDecomp(A):\n",
    "    \"\"\"A function to perform a sorted eigendecomposition a \n",
    "    an input matrix \"\"\"\n",
    "    w, v = np.linalg.eig(A)\n",
    "    idx = np.argsort(w)\n",
    "    w = w[idx]\n",
    "    v = v[:,idx]\n",
    "    return w, v\n",
    "\n",
    "    \n",
    "for runId, A in matrices.items():\n",
    "    \n",
    "    # unwieghted adjacency\n",
    "    Abinary = np.where(A > 0, 1, 0)\n",
    "    \n",
    "    # handle missing diagonal entries\n",
    "    np.fill_diagonal(Abinary, 1)\n",
    "    \n",
    "    u, s, vh = np.linalg.svd(Abinary)   \n",
    "    \n",
    "    left, width = 0.1, 0.65\n",
    "    bottom, height = 0.1, 0.65\n",
    "    spacing = 0.025\n",
    "\n",
    "    hicPlotSize = [left, bottom, width, height]\n",
    "    pcPlotSize = [left, bottom + height + spacing, width, 0.1]\n",
    "    \n",
    "    # start with a square Figure\n",
    "    fig = plt.figure(figsize=(5, 5))\n",
    "\n",
    "    ax1 = fig.add_axes(hicPlotSize)\n",
    "    ax1.imshow(Abinary, cmap='binary')\n",
    "    \n",
    "    ax2 = fig.add_axes(pcPlotSize, sharex=ax1)\n",
    "    ax2.bar(list(range(len(u))), u[:, 1], color='k')\n",
    "    ax2.set_title(runId)\n",
    "    ax2.set_xticks([])\n",
    "\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %matplotlib inline\n",
    "\n",
    "# plt.rcParams['figure.dpi'] = 200\n",
    "# # plt.rcParams['figure.figsize'] = 5, 7\n",
    "\n",
    "# for runId, A in matrices.items():\n",
    "    \n",
    "#     # binary representation\n",
    "#     Abinary = np.where(A > 0, 1, 0)\n",
    "    \n",
    "#     G = nx.from_numpy_array(A)\n",
    "#     cent = nx.eigenvector_centrality(G, max_iter=1000)\n",
    "    \n",
    "#     left, width = 0.1, 0.65\n",
    "#     bottom, height = 0.1, 0.65\n",
    "#     spacing = 0.025\n",
    "\n",
    "#     hicPlotSize = [left, bottom, width, height]\n",
    "#     centralityPlotSize = [left, bottom + height + spacing, width, 0.1]\n",
    "    \n",
    "#     # start with a square Figure\n",
    "#     fig = plt.figure(figsize=(5, 5))\n",
    "\n",
    "#     ax1 = fig.add_axes(hicPlotSize)\n",
    "#     ax1.imshow(Abinary, cmap='binary')\n",
    "    \n",
    "#     ax2 = fig.add_axes(centralityPlotSize, sharex=ax1)\n",
    "#     ax2.bar(cent.keys(), cent.values(), color='k')\n",
    "#     ax2.set_title(runId)\n",
    "#     ax2.set_xticks([])\n",
    "\n",
    "#     plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Psuedobulk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "psuedoBulk = np.zeros((matrices['run01'].shape))\n",
    "\n",
    "for runId, A in matrices.items():\n",
    "    psuedoBulk = psuedoBulk + A\n",
    "    \n",
    "\n",
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 5, 5\n",
    "\n",
    "binaryBulk = np.where(psuedoBulk > 0, 1, 0)\n",
    "\n",
    "# PCA\n",
    "pca = PCA(n_components=1, svd_solver='full')\n",
    "pca.fit(binaryBulk)\n",
    "\n",
    "pc1 = pca.components_[0, :]\n",
    "\n",
    "\n",
    "left, width = 0.1, 0.65\n",
    "bottom, height = 0.1, 0.65\n",
    "spacing = 0.025\n",
    "\n",
    "\n",
    "hicPlotSize = [left, bottom, width, height]\n",
    "pcPlotSize = [left, bottom + height + spacing, width, 0.1]\n",
    "\n",
    "# start with a square Figure\n",
    "fig = plt.figure()\n",
    "\n",
    "ax1 = fig.add_axes(hicPlotSize)\n",
    "ax1.imshow(binaryBulk, cmap='binary')\n",
    "\n",
    "\n",
    "ax2 = fig.add_axes(pcPlotSize, sharex=ax1)\n",
    "ax2.bar(list(range(len(pc1))), pc1, color='k')\n",
    "ax2.set_title('psuedo bulk')\n",
    "ax2.set_xticks([])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# plt.rcParams['figure.dpi'] = 200\n",
    "# plt.rcParams['figure.figsize'] = 5, 5\n",
    "\n",
    "# for runId, A in matrices.items():\n",
    "    \n",
    "#     plt.imshow(binaryBulk, \n",
    "#                cmap='binary', \n",
    "#                interpolation='none',\n",
    "#                alpha=0.4)\n",
    "    \n",
    "#     binaryA = np.where(A > 0, 1, 0)\n",
    "#     plt.imshow(binaryA, \n",
    "#                cmap='Reds', \n",
    "#                interpolation='none', \n",
    "#                alpha=0.6)\n",
    "    \n",
    "#     plt.title(runId)\n",
    "#     plt.show()\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Centrality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# centrality analysis\n",
    "\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 150\n",
    "plt.rcParams['figure.figsize'] = 10, 3\n",
    "\n",
    "ef = pd.DataFrame(index=range(len(chromBins)))\n",
    "\n",
    "\n",
    "for runId, A in matrices.items():\n",
    "    G = nx.from_numpy_array(A)\n",
    "    cent = nx.eigenvector_centrality(G, max_iter=1000)\n",
    "    \n",
    "    tmp = pd.Series(cent.values())\n",
    "    tmp.name = runId\n",
    "    ef = pd.concat([ef, tmp], axis=1)\n",
    "\n",
    "    \n",
    "ef.loc[:,'sum'] = ef.sum(numeric_only=True, axis=1)\n",
    "    \n",
    "print(f\"{ef.shape=}\")\n",
    "ef.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# for c in ef.columns:\n",
    "#     if not c == 'sum':\n",
    "#         plt.bar(ef.index, ef['sum'])\n",
    "#         plt.bar(ef.index, ef[c])\n",
    "#         plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 7, 7\n",
    "\n",
    "corrMat = ef[matrices.keys()].corr()\n",
    "\n",
    "sns.heatmap(corrMat, \n",
    "            cmap='viridis',  \n",
    "            center=0,\n",
    "            annot=True, \n",
    "            fmt='.2g',\n",
    "            linewidths=.5)\n",
    "\n",
    "plt.title(\"Cell vs. Cell Correlations (Eigenvector Centrality)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 9, 9\n",
    "\n",
    "corrMat = ef[matrices.keys()].T.corr()\n",
    "\n",
    "sns.heatmap(corrMat, \n",
    "            cmap='Reds',\n",
    "            alpha=0.7,\n",
    "            center=0)\n",
    "\n",
    "plt.title(\"Bin vs. Bin Correlations (Eigenvector Centrality)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # edit distance between single cells\n",
    "\n",
    "# Gsim = np.zeros((len(matrices), len(matrices)))\n",
    "\n",
    "# for i, (runId1, A1) in enumerate(matrices.items()):\n",
    "#     print(f\"{runId1=}\")\n",
    "#     for j, (runId2, A2) in enumerate(matrices.items()):\n",
    "#         print(f\"{runId2=}\")\n",
    "        \n",
    "#         G1 = nx.from_numpy_array(A1)\n",
    "#         G2 = nx.from_numpy_array(A2)\n",
    "        \n",
    "#         measure = nx.graph_edit_distance(G1, G2)\n",
    "#         Gsim[i, j] = measure\n",
    "#         Gsim[j, i] = measure\n",
    "        \n",
    "\n",
    "# plt.matshow(Gsim)      \n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for runId, A in matrices.items():\n",
    "    \n",
    "    # remove self-loops for cleaner graph\n",
    "    A = A - np.diag(np.diag(A))\n",
    "    \n",
    "    G = nx.from_numpy_array(A)\n",
    "    \n",
    "    print(\"-------------------------\")\n",
    "    print(f\"{runId=}\")\n",
    "    print(f\"{nx.average_clustering(G)=:.3f}\")\n",
    "    print(f\"{nx.density(G)=:.3f}\")\n",
    "    print(f\"{nx.number_of_edges(G)=}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "threshold = 0.5\n",
    "\n",
    "imputedContacts = []\n",
    "\n",
    "for runId, A in matrices.items():\n",
    "    \n",
    "    G = nx.from_numpy_array(A)    \n",
    "    P = np.zeros(A.shape)\n",
    "    \n",
    "#     pLink = nx.preferential_attachment(G)   \n",
    "    pLink = nx.common_neighbor_centrality(G, alpha=0.1)\n",
    "    \n",
    "    for i, j, p in pLink:\n",
    "        P[i, j] = P[i, j] + p\n",
    "        P[j, i] = P[j, i] + p\n",
    "    \n",
    "    # normalize the link scores\n",
    "    P = P / P.max()\n",
    "    \n",
    "    # threshold based prediction\n",
    "    P = np.where(P >= threshold, 1, 0)\n",
    "    \n",
    "    # impute new contacts\n",
    "    imputedA = A + P\n",
    "    imputedContacts.append(imputedA)\n",
    "    \n",
    "    fig, ax = plt.subplots(1, 3) \n",
    "    \n",
    "    ax[0].matshow(A)\n",
    "    ax[0].set_title('Original Contacts', y=1.1)\n",
    "    \n",
    "    ax[1].matshow(P)\n",
    "    ax[1].set_title('Predicted Contacts', y=1.1)\n",
    "    \n",
    "    ax[2].matshow(imputedA)\n",
    "    ax[2].set_title('Imputed Contacts', y=1.1)\n",
    "    \n",
    "    plt.suptitle(runId, y=0.7)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "imputedBulk = np.zeros((matrices['run01'].shape))\n",
    "\n",
    "for imp in imputedContacts:\n",
    "    imputedBulk += imp\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 5, 5\n",
    "\n",
    "plt.matshow(imputedBulk, cmap='plasma')\n",
    "plt.title('imputed psuedo bulk')\n",
    "plt.colorbar(fraction=0.046, pad=0.04)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 150\n",
    "plt.rcParams['figure.figsize'] = 12, 12\n",
    "plt.rcParams['figure.facecolor'] = 'white'\n",
    "\n",
    "componentThreshold = 3\n",
    "\n",
    "def removeDisconnectedComponents(G, threhsold):\n",
    "   # remove disconnected components\n",
    "    for component in list(nx.connected_components(G)):\n",
    "        if len(component) < threhsold:\n",
    "            for node in component:\n",
    "                G.remove_node(node)\n",
    "    return G\n",
    "\n",
    "                \n",
    "cmap = 'Reds'\n",
    "\n",
    "for runId, A in matrices.items():\n",
    "    \n",
    "    # remove self-loops for cleaner graph\n",
    "    A = A - np.diag(np.diag(A))\n",
    "    \n",
    "    G = nx.from_numpy_array(A)\n",
    "    \n",
    "    G = removeDisconnectedComponents(G, componentThreshold)\n",
    "    \n",
    "    degrees = [val*20 for (node, val) in G.degree()]\n",
    "    \n",
    "    G.pos = nx.circular_layout(G)\n",
    "\n",
    "    nx.draw_networkx_nodes(G, \n",
    "                           pos=G.pos,\n",
    "                           node_size=50,\n",
    "                           node_color = G.nodes,\n",
    "                           edgecolors = 'k',\n",
    "                           alpha=1,\n",
    "                           cmap=cmap)\n",
    "    \n",
    "#     nx.draw_networkx_labels(G, \n",
    "#                         pos=G.pos,\n",
    "#                         font_size=5\n",
    "#                        )\n",
    "    \n",
    "    nx.draw_networkx_edges(G,\n",
    "                       pos=G.pos,\n",
    "                       width=1,\n",
    "                       alpha=0.6)\n",
    "    \n",
    "    \n",
    "    sm = plt.cm.ScalarMappable(cmap=plt.get_cmap(cmap), \n",
    "                           norm=plt.Normalize(vmin = 0, vmax=1))\n",
    "    sm._A = []\n",
    "    cbar = plt.colorbar(sm, ticks=[0, 1], aspect=80, fraction=0.02, pad=0.012)\n",
    "    cbar.ax.set_yticklabels(['Chr Start', 'Chr End'])\n",
    "    cbar.set_label('Bin Position', rotation=270)\n",
    "    \n",
    "    \n",
    "    \n",
    "    break\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plt.rcParams['figure.dpi'] = 300\n",
    "# plt.rcParams['figure.figsize'] = 15, 15\n",
    "# plt.rcParams['figure.facecolor'] = 'white'\n",
    "\n",
    "# cmap = 'viridis'\n",
    "\n",
    "# for runId, A in matrices.items():\n",
    "    \n",
    "#     A = A - np.diag(np.diag(A))\n",
    "    \n",
    "#     print(f\"{A.max()}\")\n",
    "    \n",
    "#     G = nx.from_numpy_array(A)\n",
    "    \n",
    "#     G.pos = nx.circular_layout(G)\n",
    "    \n",
    "#     # draw the network - more control if you draw nodes and edges separately \n",
    "#     nx.draw_networkx_nodes(G,\n",
    "#                            pos=G.pos,\n",
    "#                            node_size=150,\n",
    "#                            node_color = G.nodes,\n",
    "#                            edgecolors = 'k',\n",
    "#                            alpha=0.2,\n",
    "#                            cmap=cmap\n",
    "#                           )\n",
    "    \n",
    "#     nx.draw_networkx_labels(G, \n",
    "#                             pos=G.pos,\n",
    "#                             font_size=5\n",
    "#                            )\n",
    "    \n",
    "#     nx.draw_networkx_edges(G,\n",
    "#                            pos=G.pos,\n",
    "#                            width=1,\n",
    "#                            alpha=0.5\n",
    "#                           )\n",
    "    \n",
    "\n",
    "    \n",
    "#     break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
