{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import cooler\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors\n",
    "from matplotlib.patches import Patch\n",
    "import scipy.io\n",
    "\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "from scipy import stats\n",
    "import networkx as nx\n",
    "from sklearn.metrics import jaccard_score\n",
    "\n",
    "import os\n",
    "import sys\n",
    "from importlib import reload\n",
    "\n",
    "import filters\n",
    "import pore_c_utils as pcu\n",
    "import binning_utils as binning\n",
    "import plotting as hicPlot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(pcu)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getPC1(A):\n",
    "    \"\"\"A function to get the first component\"\"\"\n",
    "    pca = PCA(n_components=1, svd_solver='full')\n",
    "    pca.fit(A)\n",
    "    return pca.components_[0, :]\n",
    "\n",
    "\n",
    "def getCorrelationMatrix(A):\n",
    "    \"\"\"A function to generate a column-wise correlation matrix\n",
    "    as described by Lieberman-Aiden 2009 \"\"\"\n",
    "    \n",
    "    C = np.zeros(A.shape)\n",
    "    n = A.shape[0]\n",
    "    \n",
    "    for i in range(n):\n",
    "        for j in range(n):\n",
    "        \n",
    "            r = stats.pearsonr(A[:, i], A[:, j])[0]\n",
    "\n",
    "            C[i, j] = r\n",
    "            C[j, i] = r\n",
    "    C = np.where(np.isnan(C), 0, C)\n",
    "    \n",
    "    return C\n",
    "\n",
    "\n",
    "def getJaccardMatrix(A):\n",
    "    \"\"\"A function to generate a column-wise Jaccard matrix \"\"\"\n",
    "    C = np.zeros(A.shape)\n",
    "    n = A.shape[0]\n",
    "    \n",
    "    for i in range(n):\n",
    "        for j in range(n):\n",
    "        \n",
    "            r = jaccard_score(A[:, i], A[:, j])\n",
    "\n",
    "            C[i, j] = r\n",
    "            C[j, i] = r\n",
    "    C = np.where(np.isnan(C), 0, C)\n",
    "    \n",
    "    return C\n",
    "        \n",
    "\n",
    "def pageRankNormalize(A, d=0.85):\n",
    "    N = A.shape[1]\n",
    "    A_hat = (d * A + (1 - d) / N)\n",
    "    return A_hat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assembly = pcu.loadAssembly(\"GRCm39_Assembly.txt\")\n",
    "\n",
    "chromDict = dict(zip(assembly['RefSeq accession'],assembly['Chromosome'].apply(lambda x : f\"chr{x}\")))\n",
    "\n",
    "# the reverse for easy lookup\n",
    "chromDict_r = {v: k for k, v in chromDict.items()}\n",
    "\n",
    "chromOrder = list(chromDict.values())\n",
    "chromOrder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reload(pcu)\n",
    "\n",
    "# load pore c data \n",
    "scPorecFilepath = \"/nfs/turbo/umms-indikar/shared/tools/Pore-C-Snakemake/results_ES5to12/matrix/\"\n",
    "\n",
    "cellNums = [1, 2, 4, 5, 6, 7] # dropping cell 3 due to sparsity\n",
    "filetags = [f\"run0{x}\" for x in cellNums]\n",
    "\n",
    "porec = {}\n",
    "\n",
    "for cell in filetags:\n",
    "    fullPath = f\"{scPorecFilepath}NlaIII_{cell}_GRCm39_unphased.matrix.mcool\"\n",
    "    \n",
    "    A, porecChromInfo, porecIndex = pcu.loadPorecCooler(fullPath, assembly, chromOrder)    \n",
    "    \n",
    "    # reindex\n",
    "    porecMat = A[porecIndex, :][:, porecIndex]\n",
    "    \n",
    "    print(f\"{cell=} {porecMat.shape=}\")\n",
    "    \n",
    "    porec[cell] = porecMat\n",
    "    \n",
    "porec.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scHICfilepath = f\"/nfs/turbo/umms-indikar/shared/projects/poreC/nagano_2017_data/nagano_1MB_raw_BALANCED.scool\"\n",
    "cellList = list(cooler.fileops.list_coolers(scHICfilepath))\n",
    "print(f\"{len(cellList)=}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sampleSize = len(porec) # sample 6 randomly\n",
    "sample = np.random.choice(cellList, sampleSize, replace=False)\n",
    "\n",
    "hic = {}\n",
    "\n",
    "for s in sample:\n",
    "    filetag = \"_\".join(s.split(\"_\")[0:2]).replace(\"/\", \"\")    \n",
    "    A, hicChromInfo, hicIndex = pcu.loadNagano2017SingleCell(scHICfilepath, s, chromOrder, balance=False)\n",
    "    \n",
    "    # reindex the matrix\n",
    "    hicMat = A[hicIndex, :][:, hicIndex]\n",
    "    print(f\"{filetag=} {hicMat.shape=}\")\n",
    "    \n",
    "    hic[filetag] = hicMat\n",
    "\n",
    "hic.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hicChromInfo.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def binarize(A):\n",
    "    return np.where(A > 0, 1, 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pseduobulk summaries\n",
    "\n",
    "def getStats(A):\n",
    "    sumContacts = np.triu(A).sum()\n",
    "    sumDiag = np.diag(A).sum()\n",
    "    sumUpper = np.triu(A).sum() - sumDiag\n",
    "    diagUpperRatio = sumDiag / sumUpper\n",
    "    \n",
    "    print(f\"{sumContacts=}\")\n",
    "    print(f\"{sumDiag=}\")\n",
    "    print(f\"{sumUpper=}\")\n",
    "    print(f\"{diagUpperRatio=:.3f}\")\n",
    "\n",
    "bulkPoreC = np.zeros(porecMat.shape)\n",
    "for runId, A in porec.items():\n",
    "    bulkPoreC += A\n",
    "    \n",
    "bulkP = binarize(bulkPoreC)\n",
    "print(\"pore-c\")\n",
    "getStats(bulkP)\n",
    "    \n",
    "bulkhiC = np.zeros(hicMat.shape)\n",
    "for runId, A in hic.items():\n",
    "    bulkhiC += A\n",
    "    \n",
    "bulkH = binarize(bulkhiC)\n",
    "print(\"\\nhi-c\")\n",
    "getStats(bulkH)\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 15, 15\n",
    "fig, (ax1, ax2) = plt.subplots(1, 2)\n",
    "\n",
    "ax1.matshow(bulkP, cmap='binary')\n",
    "ax1.set_title('Pseudobulk Pore-C (raw)')\n",
    "for idx, row in porecChromInfo.iterrows():\n",
    "    ax1.axhline(y=row['end'], lw=0.5, ls='--', c='r')\n",
    "    ax1.axvline(x=row['end'], lw=0.5, ls='--', c='r')\n",
    "\n",
    "\n",
    "ax2.matshow(bulkH, cmap='binary')\n",
    "ax2.set_title('Pseudobulk Hi-C (raw)')\n",
    "\n",
    "for idx, row in hicChromInfo.iterrows():\n",
    "    if not 'random' in row['region'] and 'M' not in row['region']:\n",
    "        \n",
    "        # remap to shuffled chrom positions\n",
    "        pos = hicIndex.index(row['end'] - 1)\n",
    "\n",
    "        ax2.axhline(y=pos, lw=0.5, ls='--', c='r')\n",
    "        ax2.axvline(x=pos, lw=0.5, ls='--', c='r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pseduobulk oberserved over expected\n",
    "    \n",
    "\n",
    "chromosome = ['chr2']\n",
    "chrHicInd = pcu.getIndices(hicChromInfo, hicIndex, chromosome, 'region')\n",
    "chrPorecInd = pcu.getIndices(porecChromInfo, porecIndex, chromosome, 'chromName')\n",
    "\n",
    "hicChr = bulkH[chrHicInd, :][:, chrHicInd]\n",
    "porecChr = bulkP[chrPorecInd, :][:, chrPorecInd]\n",
    "\n",
    "# get correlation matrices\n",
    "hiCCorr = getCorrelationMatrix(hicChr)\n",
    "porecCorr =  getCorrelationMatrix(porecChr)\n",
    "\n",
    "print(f\"{hicChr.shape=}\")\n",
    "print(f\"{porecChr.shape=}\")\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 15, 15\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "\n",
    "fig.add_subplot(221)\n",
    "plt.imshow(porecChr, cmap='binary')\n",
    "plt.title(f'Ensemble Pore-C (raw, {chromosome})')\n",
    "\n",
    "fig.add_subplot(222)\n",
    "plt.imshow(hicChr, cmap='binary')\n",
    "plt.title(f'Ensemble Hi-C (raw, {chromosome})')\n",
    "\n",
    "# fig.add_subplot(223)\n",
    "# plt.imshow(porecCorr, cmap='binary')\n",
    "# plt.title(f'Ensemble Pore-C Correlation Matrix (raw, {chromosome})')\n",
    "\n",
    "# fig.add_subplot(224)\n",
    "# plt.imshow(hiCCorr, cmap='binary')\n",
    "# plt.title(f'Ensemble Hi-C Correlation Matrix (raw, {chromosome})')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pseduobulk oberserved over expected\n",
    "chromosome = ['chr2', 'chr5', 'chr13']\n",
    "chrHicInd = pcu.getIndices(hicChromInfo, hicIndex, chromosome, 'region')\n",
    "chrPorecInd = pcu.getIndices(porecChromInfo, porecIndex, chromosome, 'chromName')\n",
    "\n",
    "hicChr = bulkH[chrHicInd, :][:, chrHicInd]\n",
    "porecChr = bulkP[chrPorecInd, :][:, chrPorecInd]\n",
    "\n",
    "hicChr = pcu.dropZeroRows(hicChr)\n",
    "porecChr = pcu.dropZeroRows(porecChr)\n",
    "\n",
    "\n",
    "print(f\"{hicChr.shape=}\")\n",
    "print(f\"{porecChr.shape=}\")\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 15, 15\n",
    "fig, (ax1, ax2) = plt.subplots(1, 2)\n",
    "\n",
    "ax1.matshow(porecChr, cmap='binary')\n",
    "ax1.set_title(f'Pseudobulk Pore-C (raw, {chromosome})')\n",
    "\n",
    "cnt = 0\n",
    "for chrm in chromosome[:-1]:    \n",
    "    row = porecChromInfo.loc[porecChromInfo['chromName'] == chrm]\n",
    "    # remap to shuffled chrom position\n",
    "    \n",
    "    pos = row['length'].values[0] - 1\n",
    "    cnt += pos\n",
    "    \n",
    "    ax1.axhline(y=cnt, lw=1,  c='r')    \n",
    "    ax1.axvline(x=cnt, lw=1,  c='r')  \n",
    "    \n",
    "\n",
    "ax2.matshow(hicChr, cmap='binary')\n",
    "ax2.set_title(f'Pseudobulk Hi-C (raw, {chromosome})')\n",
    "\n",
    "\n",
    "cnt = 0\n",
    "for chrm in chromosome[:-1]:    \n",
    "    row = hicChromInfo.loc[hicChromInfo['region'] == chrm]\n",
    "    # remap to shuffled chrom position\n",
    "    \n",
    "    pos = row['length'].values[0] - 1\n",
    "    cnt += pos\n",
    "    \n",
    "    ax2.axhline(y=cnt, lw=1,  c='r')    \n",
    "    ax2.axvline(x=cnt, lw=1,  c='r')  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filtered scPore-C Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the filtered pore-C data\n",
    "dirpath = \"/nfs/turbo/umms-indikar/shared/tools/Pore-C-Snakemake/results_ES5to12/contacts/\"\n",
    "cellNums = [1, 2, 4, 5, 6, 7] # dropping cell 3 due to sparsity\n",
    "filetags = [f\"run0{x}\" for x in cellNums]\n",
    "\n",
    "# filetags = [f\"run0{x}\" for x in range(1, 2)]\n",
    "\n",
    "assembly = pcu.loadAssembly(\"GRCm39_Assembly.txt\")\n",
    "rawCells = pcu.loadRawContactDirectory(dirpath, filetags)\n",
    "\n",
    "print(\"Filtering....\")\n",
    "filteredCells = filters.runAllFilters(rawCells,\n",
    "                              assembly,\n",
    "                              adjacent=False,\n",
    "                              chromosome=None, \n",
    "                              verbose=False)  \n",
    "\n",
    "print(f\"\\n{filteredCells.keys()=}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "resolution = 1000000 # 1mb scale\n",
    "\n",
    "totalLength = assembly['chromEnd'].max()\n",
    "bins = binning.getBins(totalLength, resolution)\n",
    "\n",
    "label1 = 'bin1'\n",
    "label2 = 'bin2'\n",
    "\n",
    "binnedData = {}\n",
    "matrices = {}\n",
    "\n",
    "for runId, data in filteredCells.items():\n",
    "    binned = binning.wholeGenomeBinData(data, bins, label1, label2)\n",
    "    binnedData[runId] = binned\n",
    "      \n",
    "    A = pcu.getHic(binned, bins, label1, label2)\n",
    "    print(f\"{runId=} {A.shape=}\")\n",
    "    matrices[runId] = A\n",
    "    \n",
    "print(f\"{matrices.keys()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assembly.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "newRows = []\n",
    "\n",
    "for idx, row in assembly.iterrows():\n",
    "    if \"NC\" in row['RefSeq accession']:\n",
    "        \n",
    "        start = binning.returnBin(row['chromStart'], bins) + 2\n",
    "        end = binning.returnBin(row['chromEnd'], bins) + 3\n",
    "        \n",
    "        newRow = {\n",
    "            'region' : f\"chr{row['Chromosome']}\",\n",
    "            'start' : start,\n",
    "            'end' : end,\n",
    "            'length' : end - start,\n",
    "        }\n",
    "        \n",
    "        newRows.append(newRow)\n",
    "        \n",
    "filteredChromInfo = pd.DataFrame(newRows)\n",
    "filteredChromInfo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# individual cell summaries\n",
    "\n",
    "bulkPoreCFiltered = np.zeros(A.shape)\n",
    "for runId, A in matrices.items():\n",
    "    bulkPoreCFiltered += A\n",
    "    \n",
    "bulkPFiltered = binarize(bulkPoreCFiltered)\n",
    "    \n",
    "print(\"filtered pore-c\")\n",
    "getStats(bulkPFiltered)\n",
    "    \n",
    "print(\"\\nhi-c\")\n",
    "getStats(bulkH)\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 15, 15\n",
    "fig, (ax1, ax2) = plt.subplots(1, 2)\n",
    "\n",
    "ax1.matshow(bulkPFiltered, cmap='binary')\n",
    "ax1.set_title('Pseudobulk Pore-C (raw)')\n",
    "for idx, row in filteredChromInfo.iterrows():\n",
    "    ax1.axhline(y=row['end'], lw=0.5, ls='--', c='r')\n",
    "    ax1.axvline(x=row['end'], lw=0.5, ls='--', c='r')\n",
    "\n",
    "\n",
    "ax2.matshow(bulkH, cmap='binary')\n",
    "ax2.set_title('Pseudobulk Hi-C (unfiltered)')\n",
    "\n",
    "for idx, row in hicChromInfo.iterrows():\n",
    "    if not 'random' in row['region'] and 'M' not in row['region']:\n",
    "        \n",
    "        # remap to shuffled chrom positions\n",
    "        pos = hicIndex.index(row['end'] - 1)\n",
    "\n",
    "        ax2.axhline(y=pos, lw=0.5, ls='--', c='r')\n",
    "        ax2.axvline(x=pos, lw=0.5, ls='--', c='r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filteredPoreCIndex = list(range(0, len(bulkPFiltered)))\n",
    "print(f\"{len(filteredPoreCIndex)=}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosome = ['chr2']\n",
    "chrHicInd = pcu.getIndices(hicChromInfo, hicIndex, chromosome, 'region')\n",
    "chrPorecInd = pcu.getIndices(filteredChromInfo, filteredPoreCIndex, chromosome, 'region')\n",
    "\n",
    "hicChr = bulkH[chrHicInd, :][:, chrHicInd]\n",
    "porecChr = bulkPFiltered[chrPorecInd, :][:, chrPorecInd]\n",
    "\n",
    "# get correlation matrices\n",
    "hiCCorr = getCorrelationMatrix(hicChr)\n",
    "porecCorr =  getCorrelationMatrix(porecChr)\n",
    "\n",
    "print(f\"{hicChr.shape=}\")\n",
    "print(f\"{porecChr.shape=}\")\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 15, 15\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "\n",
    "fig.add_subplot(221)\n",
    "plt.imshow(porecChr, cmap='binary')\n",
    "plt.title(f'Ensemble Pore-C (filtered, {chromosome})')\n",
    "\n",
    "fig.add_subplot(222)\n",
    "plt.imshow(hicChr, cmap='binary')\n",
    "plt.title(f'Ensemble Hi-C (raw, {chromosome})')\n",
    "\n",
    "# fig.add_subplot(223)\n",
    "# plt.imshow(porecCorr, cmap='binary')\n",
    "# plt.title(f'Ensemble Pore-C Correlation Matrix (filtered, {chromosome})')\n",
    "\n",
    "# fig.add_subplot(224)\n",
    "# plt.imshow(hiCCorr, cmap='binary')\n",
    "# plt.title(f'Ensemble Hi-C Correlation Matrix (raw, {chromosome})')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # save a single .mat file\n",
    "\n",
    "# outputdir = \"data/\"\n",
    "\n",
    "# chromosome = ['chr2']\n",
    "\n",
    "# assert(len(chromosome) == 1)\n",
    "\n",
    "# for pKey in matrices.keys():\n",
    "    \n",
    "#     print(f\"{pKey=}\")\n",
    "    \n",
    "#     # get matrices and filter\n",
    "#     pMat = matrices[pKey]\n",
    "        \n",
    "#     chrPorecInd = getIndices(filteredChromInfo, filteredPoreCIndex, chromosome, 'region')\n",
    "    \n",
    "#     A = pMat[:, chrPorecInd][chrPorecInd, :]\n",
    "#     A = np.where(A > 0, 1, 0)\n",
    "    \n",
    "#     fileName = f\"{outputdir}{pKey}{chromosome[0]}.mat\"\n",
    "    \n",
    "#     scipy.io.savemat(fileName, {'A': A})\n",
    "    \n",
    "    \n",
    "# print('done.')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pseduobulk \n",
    "chromosome = ['chr2', 'chr5', 'chr13']\n",
    "chrHicInd = pcu.getIndices(hicChromInfo, hicIndex, chromosome, 'region')\n",
    "chrPorecInd = pcu.getIndices(filteredChromInfo, filteredPoreCIndex, chromosome, 'region')\n",
    "\n",
    "hicChr = bulkH[chrHicInd, :][:, chrHicInd]\n",
    "porecChr = bulkPFiltered[chrPorecInd, :][:, chrPorecInd]\n",
    "\n",
    "print(f\"{hicChr.shape=}\")\n",
    "print(f\"{porecChr.shape=}\")\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 15, 15\n",
    "fig, (ax1, ax2) = plt.subplots(1, 2)\n",
    "\n",
    "ax1.matshow(porecChr, cmap='binary')\n",
    "ax1.set_title(f'Pseudobulk Pore-C (filtered, {chromosome})')\n",
    "\n",
    "cnt = 0\n",
    "for chrm in chromosome[:-1]:    \n",
    "    row = filteredChromInfo.loc[filteredChromInfo['region'] == chrm]\n",
    "    # remap to shuffled chrom position\n",
    "    \n",
    "    pos = row['length'].values[0] - 1\n",
    "    cnt += pos\n",
    "    \n",
    "    ax1.axhline(y=cnt, lw=1,  c='r')    \n",
    "    ax1.axvline(x=cnt, lw=1,  c='r')  \n",
    "    \n",
    "\n",
    "ax2.matshow(hicChr, cmap='binary')\n",
    "ax2.set_title(f'Pseudobulk Hi-C (raw, {chromosome})')\n",
    "\n",
    "\n",
    "cnt = 0\n",
    "for chrm in chromosome[:-1]:    \n",
    "    row = hicChromInfo.loc[hicChromInfo['region'] == chrm]\n",
    "    # remap to shuffled chrom position\n",
    "    \n",
    "    pos = row['length'].values[0] - 1\n",
    "    cnt += pos\n",
    "    \n",
    "    ax2.axhline(y=cnt, lw=1,  c='r')    \n",
    "    ax2.axvline(x=cnt, lw=1,  c='r')  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosome = ['chr2']\n",
    "chrHicInd = pcu.getIndices(hicChromInfo, hicIndex, chromosome, 'region')\n",
    "chrPorecInd = pcu.getIndices(filteredChromInfo, filteredPoreCIndex, chromosome, 'region')\n",
    "\n",
    "hMat = bulkH[chrHicInd, :][:, chrHicInd]\n",
    "pMat = bulkPFiltered[chrPorecInd, :][:, chrPorecInd]\n",
    "\n",
    "np.fill_diagonal(pMat, 1)\n",
    "np.fill_diagonal(hMat, 1)\n",
    "\n",
    "hMat = pcu.forceAdjacentConnections(hMat)\n",
    "pMat = pcu.forceAdjacentConnections(pMat)\n",
    "\n",
    "pC = getCorrelationMatrix(pMat)\n",
    "hC = getCorrelationMatrix(hMat)\n",
    "\n",
    "porecPC1 = getPC1(pC)\n",
    "hicPC1 = getPC1(hC)\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 10, 4\n",
    "fig = plt.figure()\n",
    "\n",
    "fig.add_subplot(221)\n",
    "plt.bar(range(len(porecPC1)), porecPC1, color='k')\n",
    "\n",
    "fig.add_subplot(222)\n",
    "plt.bar(range(len(hicPC1)), hicPC1, color='k')\n",
    "\n",
    "fig.add_subplot(223)\n",
    "plt.imshow(pC, cmap='binary')\n",
    "\n",
    "fig.add_subplot(224)\n",
    "plt.imshow(hC, cmap='binary')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# u, s, vh = np.linalg.svd(hC)\n",
    "\n",
    "# plt.bar(range(len(s)), -u[:, 0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Individual Cell Comparisons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosome = ['chr2']\n",
    "\n",
    "zippedKeys = list(zip(matrices.keys(), hic.keys()))\n",
    "for pKey, hKey in zippedKeys:\n",
    "    \n",
    "    print(f\"{pKey=} {hKey=}\")\n",
    "    \n",
    "    \n",
    "    # get matrices and filter\n",
    "    pMat = matrices[pKey]\n",
    "    hMat = hic[hKey]\n",
    "        \n",
    "    chrPorecInd = pcu.getIndices(filteredChromInfo, filteredPoreCIndex, chromosome, 'region')\n",
    "    chrHicInd = pcu.getIndices(hicChromInfo, hicIndex, chromosome, 'region')\n",
    "    \n",
    "    pMat = binarize(pMat[:, chrPorecInd][chrPorecInd, :])\n",
    "    hMat = binarize(hMat[:, chrHicInd][chrHicInd, :])\n",
    "    \n",
    "    print(f\"{pMat.shape=} {hMat.shape=}\")\n",
    "        \n",
    "    plt.rcParams['figure.dpi'] = 200\n",
    "    plt.rcParams['figure.figsize'] = 15, 15\n",
    "    fig, (ax1, ax2) = plt.subplots(1, 2)\n",
    "    \n",
    "    ax1.matshow(pMat, cmap='binary')\n",
    "    ax1.set_title(f'scPore-C - {pKey} (f{chromosome})')\n",
    "\n",
    "    cnt = 0\n",
    "    for chrm in chromosome[:-1]:    \n",
    "        row = filteredChromInfo.loc[filteredChromInfo['region'] == chrm]\n",
    "        # remap to shuffled chrom position\n",
    "\n",
    "        pos = row['length'].values[0] - 1\n",
    "        cnt += pos\n",
    "\n",
    "        ax1.axhline(y=cnt, lw=2,  c='r')    \n",
    "        ax1.axvline(x=cnt, lw=2,  c='r')  \n",
    "\n",
    "\n",
    "    ax2.matshow(hMat, cmap='binary')\n",
    "    ax2.set_title(f'scHi-C - {hKey} ({chromosome})')\n",
    "\n",
    "\n",
    "    cnt = 0\n",
    "    for chrm in chromosome[:-1]:    \n",
    "        row = hicChromInfo.loc[hicChromInfo['region'] == chrm]\n",
    "        # remap to shuffled chrom position\n",
    "\n",
    "        pos = row['length'].values[0] - 1\n",
    "        cnt += pos\n",
    "\n",
    "        ax2.axhline(y=cnt, lw=2,  c='r')    \n",
    "        ax2.axvline(x=cnt, lw=2,  c='r')  \n",
    "    \n",
    "    plt.show()\n",
    "    \n",
    "    break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosome = ['chr2']\n",
    "\n",
    "zippedKeys = list(zip(matrices.keys(), hic.keys()))\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 200\n",
    "plt.rcParams['figure.figsize'] = 20, 10\n",
    "fig = plt.figure()\n",
    "    \n",
    "count = 0\n",
    "for pKey, hKey in zippedKeys:\n",
    "    print(\"----------------\")\n",
    "    print(f\"{pKey=} {hKey=}\")\n",
    "    \n",
    "    # get matrices and filter\n",
    "    pMat = matrices[pKey]\n",
    "    hMat = hic[hKey]\n",
    "        \n",
    "    chrPorecInd = pcu.getIndices(filteredChromInfo, filteredPoreCIndex, chromosome, 'region')\n",
    "    chrHicInd = pcu.getIndices(hicChromInfo, hicIndex, chromosome, 'region')\n",
    "    \n",
    "    pMat = binarize(pMat[:, chrPorecInd][chrPorecInd, :])\n",
    "    hMat = binarize(hMat[:, chrHicInd][chrHicInd, :])\n",
    "    \n",
    "    print(f\"{pMat.shape=} {hMat.shape=}\")\n",
    "    \n",
    "    pRowSum = np.sum(pMat, axis=0)\n",
    "    hRowSum = np.sum(hMat, axis=0)\n",
    "    \n",
    "    fig.add_subplot(len(matrices), 2, count+1)    \n",
    "    plt.bar(range(len(pRowSum)), pRowSum)\n",
    "    plt.title(f\"{pKey}\")\n",
    "    \n",
    "    fig.add_subplot(len(matrices), 2, count+2)    \n",
    "    plt.bar(range(len(hRowSum)), hRowSum)\n",
    "    plt.title(f\"{hKey}\")\n",
    "\n",
    "    count += 2\n",
    "    \n",
    "plt.subplots_adjust(left=0.1,\n",
    "                    bottom=0.1, \n",
    "                    right=0.9, \n",
    "                    top=0.9, \n",
    "                    wspace=0.1, \n",
    "                    hspace=0.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# trans Conparisons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosome = ['chr2']\n",
    "colIndex = pcu.getIndices(filteredChromInfo, filteredPoreCIndex, chromosome, 'region')\n",
    "\n",
    "\n",
    "plt.rcParams['figure.figsize'] = 4, 13\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "figCount = 0\n",
    "for runId, A in matrices.items():\n",
    "    figCount += 1 \n",
    "    \n",
    "    transA = A[:, colIndex]\n",
    "    \n",
    "    transA = np.where(transA > 0, 1, 0)\n",
    "    \n",
    "    fig.add_subplot(1, len(matrices), figCount)\n",
    "    \n",
    "    plt.spy(transA, marker='s', c='b', markersize=0.5, alpha=0.4)\n",
    "    \n",
    "    cnt = 0\n",
    "    \n",
    "    ytick = []\n",
    "    ytickLabels = []\n",
    "    \n",
    "    for chrm in chromOrder:    \n",
    "        row = filteredChromInfo.loc[filteredChromInfo['region'] == chrm]\n",
    "        # remap to shuffled chrom position\n",
    "\n",
    "        pos = row['length'].values[0] - 1\n",
    "        cnt += pos\n",
    "        \n",
    "        ytick.append(cnt - pos // 2)\n",
    "        ytickLabels.append(chrm)\n",
    "\n",
    "        plt.axhline(y=cnt, lw=0.5, ls='--',  c='k')    \n",
    "        \n",
    "    plt.xticks([], [])\n",
    "    plt.title(f\"{runId}\", y=1.001)\n",
    "    if figCount == 1:\n",
    "        plt.yticks(ytick, ytickLabels)\n",
    "    else:\n",
    "        plt.yticks([], [])\n",
    "\n",
    "plt.subplots_adjust(left=0.01,\n",
    "                    bottom=0.1, \n",
    "                    right=0.99, \n",
    "                    top=0.9, \n",
    "                    wspace=0.2, \n",
    "                    hspace=0.01)    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PC1 Comparisons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosome = ['chr2']\n",
    "\n",
    "zippedKeys = list(zip(matrices.keys(), hic.keys()))\n",
    "for pKey, hKey in zippedKeys:\n",
    "    \n",
    "    print(f\"\\n{pKey=} {hKey=}\")\n",
    "    \n",
    "    \n",
    "    # get matrices and filter\n",
    "    pMat = matrices[pKey]\n",
    "    hMat = hic[hKey]\n",
    "        \n",
    "    chrPorecInd = pcu.getIndices(filteredChromInfo, filteredPoreCIndex, chromosome, 'region')\n",
    "    chrHicInd = pcu.getIndices(hicChromInfo, hicIndex, chromosome, 'region')\n",
    "    \n",
    "    pMat = binarize(pMat[:, chrPorecInd][chrPorecInd, :])\n",
    "    hMat = binarize(hMat[:, chrHicInd][chrHicInd, :])\n",
    "    \n",
    "    print(f\"{pMat.shape=} {hMat.shape=}\")\n",
    "    \n",
    "    # fill the diagonal\n",
    "    np.fill_diagonal(pMat, 1)\n",
    "    np.fill_diagonal(hMat, 1)\n",
    "    \n",
    "    # force adjacent connections\n",
    "    hMat = pcu.forceAdjacentConnections(hMat)\n",
    "    pMat = pcu.forceAdjacentConnections(pMat)\n",
    "    \n",
    "#     pMat = pageRankNormalize(pMat)\n",
    "#     hMat = pageRankNormalize(hMat)\n",
    "    \n",
    "#     pC = getCorrelationMatrix(pMat)\n",
    "#     hC = getCorrelationMatrix(hMat)\n",
    "\n",
    "    pC = getJaccardMatrix(pMat)\n",
    "    hC = getJaccardMatrix(hMat)\n",
    "\n",
    "    \n",
    "    print(f\"{np.linalg.matrix_rank(pC)=} {np.linalg.matrix_rank(hC)=}\")\n",
    "    \n",
    "    pPC1 = getPC1(pC)\n",
    "    hPC1 = getPC1(hC)\n",
    "    \n",
    "    r = stats.pearsonr(pPC1, hPC1)\n",
    "    \n",
    "    correlation = r[0]\n",
    "    pvalue = r[1]\n",
    "    \n",
    "    print(f\"{correlation=:.4f}, {pvalue=:.4f}\")\n",
    "    \n",
    "    \n",
    "    plt.rcParams['figure.dpi'] = 200\n",
    "    plt.rcParams['figure.figsize'] = 10, 10\n",
    "    \n",
    "    left, width = 0.1, 0.65\n",
    "    bottom, height = 0.1, 0.65\n",
    "    spacing = 0.025\n",
    "\n",
    "\n",
    "    hicPlotSize = [left, bottom, width, height]\n",
    "    pcPlotSize = [left, bottom + height + spacing, width, 0.1]\n",
    "\n",
    "    # start with a square Figure\n",
    "    fig = plt.figure()\n",
    "\n",
    "    ax1 = fig.add_axes(hicPlotSize)\n",
    "    \n",
    "    c_white = matplotlib.colors.colorConverter.to_rgba('white', alpha = 0)\n",
    "    c_black = matplotlib.colors.colorConverter.to_rgba('black', alpha = 1)\n",
    "    \n",
    "    c_C0 = matplotlib.colors.colorConverter.to_rgba('C0', alpha = 0.7)\n",
    "    c_C1 = matplotlib.colors.colorConverter.to_rgba('C1', alpha = 0.7)\n",
    "\n",
    "    cmap1 = matplotlib.colors.LinearSegmentedColormap.from_list('cmap1',[c_white, c_black], 512)\n",
    "    cmap2 = matplotlib.colors.LinearSegmentedColormap.from_list('cmap2',[c_white, c_C0], 512)\n",
    "    cmap3 = matplotlib.colors.LinearSegmentedColormap.from_list('cmap3',[c_white, c_C1], 512)\n",
    "    \n",
    "    \n",
    "    T = hMat + pMat\n",
    "    commmonContacts = np.where(T == 2, 1, 0)\n",
    "    \n",
    "    ax1.imshow(pMat, cmap=cmap2)\n",
    "    ax1.imshow(hMat, cmap=cmap3)    \n",
    "    ax1.imshow(commmonContacts, cmap=cmap1)\n",
    "    \n",
    "    \n",
    "    legend_elements = [Patch(facecolor='k', edgecolor='k', label='Common Contacts'),\n",
    "                       Patch(facecolor=c_C0, edgecolor='k', label=pKey),\n",
    "                       Patch(facecolor=c_C1, edgecolor='k', label=hKey)]\n",
    "    \n",
    "    ax1.legend(handles=legend_elements)\n",
    "\n",
    "    ax2 = fig.add_axes(pcPlotSize, sharex=ax1)\n",
    "    \n",
    "    ax2.bar(range(len(pPC1)), pPC1, alpha=0.8, label=pKey)    \n",
    "    ax2.bar(range(len(hPC1)), hPC1, alpha=0.5, label=hKey)\n",
    "    ax2.legend()\n",
    "    \n",
    "    ax2.set_title(f\"Single-Cell Comparison({chromosome})\")\n",
    "    \n",
    "    cnt = 0\n",
    "    for chrm in chromosome[:-1]:    \n",
    "        row = filteredChromInfo.loc[filteredChromInfo['region'] == chrm]\n",
    "        # remap to shuffled chrom position\n",
    "\n",
    "        pos = row['length'].values[0] - 1\n",
    "        cnt += pos\n",
    "\n",
    "        ax1.axhline(y=cnt, lw=1,ls=\"--\", c='r')    \n",
    "        ax1.axvline(x=cnt, lw=1,ls=\"--\", c='r')  \n",
    "        ax2.axvline(x=cnt, lw=1,ls=\"--\", c='r')  \n",
    "    \n",
    "    plt.show()\n",
    "    break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imputation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def imputeContacts(A, tau=0.5, alpha=0.1):\n",
    "    \"\"\"A functiont to impute contacts based on the existing\n",
    "    structure \"\"\"\n",
    "    G = nx.from_numpy_array(A)    \n",
    "    P = np.zeros(A.shape)\n",
    "    \n",
    "    pLink = nx.common_neighbor_centrality(G, alpha=alpha)\n",
    "    for i, j, p in pLink:\n",
    "        P[i, j] = P[i, j] + p\n",
    "        P[j, i] = P[j, i] + p\n",
    "    \n",
    "    # normalize the link scores\n",
    "    P = P / P.max()\n",
    "    \n",
    "    # threshold based prediction\n",
    "    P = np.where(P >= tau, 1, 0)\n",
    "    return P"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosome = ['chr2']\n",
    "\n",
    "alpha = 0.5\n",
    "tau = 0.6\n",
    "\n",
    "zippedKeys = list(zip(matrices.keys(), hic.keys()))\n",
    "for pKey, hKey in zippedKeys:\n",
    "    \n",
    "    print(f\"\\n{pKey=} {hKey=}\")\n",
    "    \n",
    "    \n",
    "    # get matrices and filter\n",
    "    pMat = matrices[pKey]\n",
    "    hMat = hic[hKey]\n",
    "        \n",
    "    chrPorecInd = pcu.getIndices(filteredChromInfo, filteredPoreCIndex, chromosome, 'region')\n",
    "    chrHicInd = pcu.getIndices(hicChromInfo, hicIndex, chromosome, 'region')\n",
    "    \n",
    "    pMat = binarize(pMat[:, chrPorecInd][chrPorecInd, :])\n",
    "    hMat = binarize(hMat[:, chrHicInd][chrHicInd, :])\n",
    "    \n",
    "    porecP = imputeContacts(pMat, tau, alpha)\n",
    "    hicP = imputeContacts(hMat, tau, alpha)\n",
    "    \n",
    "    plt.rcParams['figure.dpi'] = 200\n",
    "    plt.rcParams['figure.figsize'] = 15, 15\n",
    "    fig, (ax1, ax2) = plt.subplots(1, 2)\n",
    "    \n",
    "    c_white = matplotlib.colors.colorConverter.to_rgba('white',alpha = 0)\n",
    "    c_black= matplotlib.colors.colorConverter.to_rgba('black',alpha = 1)\n",
    "    cmap1 = matplotlib.colors.LinearSegmentedColormap.from_list('cmap1',[c_white,c_black],512)\n",
    "    \n",
    "    ax1.matshow(pMat, cmap=cmap1)\n",
    "    ax1.set_title(f'scPore-C - {pKey} (f{chromosome})')\n",
    "    \n",
    "    c_white = matplotlib.colors.colorConverter.to_rgba('white',alpha = 0)\n",
    "    c_blue = matplotlib.colors.colorConverter.to_rgba('blue',alpha = 0.5)\n",
    "    cmap2 = matplotlib.colors.LinearSegmentedColormap.from_list('cmap2',[c_white,c_blue],512)\n",
    "    \n",
    "    ax1.matshow(porecP, cmap=cmap2)\n",
    "    \n",
    "    cnt = 0\n",
    "    for chrm in chromosome[:-1]:    \n",
    "        row = filteredChromInfo.loc[filteredChromInfo['region'] == chrm]\n",
    "        # remap to shuffled chrom position\n",
    "\n",
    "        pos = row['length'].values[0] - 1\n",
    "        cnt += pos\n",
    "\n",
    "        ax1.axhline(y=cnt, lw=1,  c='r')    \n",
    "        ax1.axvline(x=cnt, lw=1,  c='r')  \n",
    "\n",
    "\n",
    "    c_white = matplotlib.colors.colorConverter.to_rgba('white',alpha = 0)\n",
    "    c_black= matplotlib.colors.colorConverter.to_rgba('black',alpha = 1)\n",
    "    cmap1 = matplotlib.colors.LinearSegmentedColormap.from_list('cmap1',[c_white,c_black],512)\n",
    "    \n",
    "    ax2.matshow(hMat, cmap=cmap1)\n",
    "    \n",
    "    c_white = matplotlib.colors.colorConverter.to_rgba('white',alpha = 0)\n",
    "    c_blue = matplotlib.colors.colorConverter.to_rgba('blue',alpha = 0.5)\n",
    "    cmap2 = matplotlib.colors.LinearSegmentedColormap.from_list('cmap2',[c_white,c_blue],512)\n",
    "    \n",
    "#     ax2.matshow(hicP, cmap=cmap2)\n",
    "    \n",
    "#     ax2.set_title(f'scHi-C - {hKey} ({chromosome})')\n",
    "\n",
    "\n",
    "    cnt = 0\n",
    "    for chrm in chromosome[:-1]:    \n",
    "        row = hicChromInfo.loc[hicChromInfo['region'] == chrm]\n",
    "        # remap to shuffled chrom position\n",
    "\n",
    "        pos = row['length'].values[0] - 1\n",
    "        cnt += pos\n",
    "\n",
    "        ax2.axhline(y=cnt, lw=1,  c='r')    \n",
    "        ax2.axvline(x=cnt, lw=1,  c='r')  \n",
    "    \n",
    "    plt.show()\n",
    "    \n",
    "    break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# zippedKeys = list(zip(porec.keys(), hic.keys()))\n",
    "        \n",
    "\n",
    "# for pKey, hKey in zippedKeys:\n",
    "    \n",
    "#     print(\"------------------------\")\n",
    "#     print(f\"{pKey=}\")\n",
    "#     print(f\"{hKey=}\")\n",
    "#     pMat = porec[pKey].matrix(balance=True)[:]\n",
    "#     hMat = hic[hKey].matrix(balance=True)[:]\n",
    "    \n",
    "    \n",
    "#     # reindex the data\n",
    "#     hMat = hMat[hicIndex, :][:, hicIndex]\n",
    "#     pMat = pMat[porecIndex, :][:, porecIndex]\n",
    "    \n",
    "#     print(f\"{hMat.shape=}\")\n",
    "#     print(f\"{pMat.shape=}\")\n",
    "    \n",
    "    \n",
    "#      # binarize both\n",
    "#     pMatBinary = np.where(pMat > 0, 1, 0)\n",
    "#     hMatBinary = np.where(hMat > 0, 1, 0)\n",
    "    \n",
    "#     A = np.zeros(hMat.shape)\n",
    "    \n",
    "#     # truncate the y chrome of the porec data\n",
    "#     n = hMat.shape[0]\n",
    "#     pMatBinary = pMatBinary[0:n, 0:n]\n",
    "    \n",
    "#     # get upper and lower triangles\n",
    "#     pMatTri = np.triu(pMatBinary)\n",
    "#     hMatTri = np.tril(hMatBinary)\n",
    "    \n",
    "#     A += pMatTri\n",
    "#     A += hMatTri\n",
    "    \n",
    "#     ABinary = np.where(A > 0, 1, 0)\n",
    "    \n",
    "#     plt.matshow(ABinary, cmap='binary')\n",
    "#     plt.title(f'Raw scPore-C Contacts {pKey}', y=1.1)\n",
    "#     plt.text((n//2), n+300, f\"Raw scHi-C Contacts {hKey}\", \n",
    "#              horizontalalignment='center', fontsize=12)\n",
    "    \n",
    "#     plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plt.rcParams['figure.dpi'] = 200\n",
    "# plt.rcParams['figure.figsize'] = 10, 10\n",
    "\n",
    "# zippedKeys = list(zip(poreC.keys(), hiC.keys()))\n",
    "\n",
    "# for pKey, hKey in zippedKeys:\n",
    "    \n",
    "#     print(\"------------------------\")\n",
    "#     print(f\"{pKey=}\")\n",
    "#     print(f\"{hKey=}\")\n",
    "#     pMat = poreC[pKey].matrix(balance=True)[:]\n",
    "#     hMat = hiC[hKey].matrix(balance=True)[:]\n",
    "    \n",
    "#     # binarize both\n",
    "#     pMatBinary = np.where(pMat > 0, 1, 0)\n",
    "#     hMatBinary = np.where(hMat > 0, 1, 0)\n",
    "    \n",
    "#     sumPoreC = np.triu(pMatBinary).sum()\n",
    "#     sumPoreCDiag = np.diag(pMatBinary).sum()\n",
    "    \n",
    "#     sumhiC = np.triu(hMatBinary).sum()\n",
    "#     sumhiCDiag = np.diag(hMatBinary).sum()\n",
    "    \n",
    "#     print()\n",
    "#     print(f\"{sumPoreC=}\")\n",
    "#     print(f\"{sumPoreCDiag=}\")\n",
    "#     print(f\"{sumPoreCDiag/sumPoreC=:.3f}\")\n",
    "    \n",
    "#     print()\n",
    "#     print(f\"{sumhiC=}\")\n",
    "#     print(f\"{sumhiCDiag=}\")\n",
    "#     print(f\"{sumhiCDiag/sumPoreC=:.3f}\")\n",
    "    \n",
    "\n",
    "# #     fig, ax = plt.subplots(1, 2) \n",
    "    \n",
    "# #     ax[0].matshow(pMatBinary, cmap='binary')\n",
    "# #     ax[0].set_title(f'scPore-C Contacts: {pKey}', y=1.2)\n",
    "\n",
    "# #     ax[1].matshow(hMatBinary, cmap='binary')\n",
    "# #     ax[1].set_title(f'scHi-C Contacts: {hKey}', y=1.2)\n",
    "    \n",
    "# #     plt.show()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plt.rcParams['figure.dpi'] = 200\n",
    "# plt.rcParams['figure.figsize'] = 10, 10\n",
    "\n",
    "# zippedKeys = list(zip(poreC.keys(), hiC.keys()))\n",
    "\n",
    "# for pKey, hKey in zippedKeys:\n",
    "    \n",
    "#     print(\"------------------------\")\n",
    "#     print(f\"{pKey=}\")\n",
    "#     print(f\"{hKey=}\")\n",
    "#     pMat = poreC[pKey].matrix(balance=True)[:]\n",
    "#     hMat = hiC[hKey].matrix(balance=True)[:]\n",
    "    \n",
    "#     # binarize both\n",
    "#     pMatBinary = np.where(pMat > 0, 1, 0)\n",
    "#     hMatBinary = np.where(hMat > 0, 1, 0)\n",
    "    \n",
    "#     A = np.zeros(hMat.shape)\n",
    "    \n",
    "#     # truncate the y chrome of the porec data\n",
    "#     n = hMat.shape[0]\n",
    "#     pMatBinary = pMatBinary[0:n, 0:n]\n",
    "    \n",
    "#     # get upper and lower triangles\n",
    "#     pMatTri = np.triu(pMatBinary)\n",
    "#     hMatTri = np.tril(hMatBinary)\n",
    "    \n",
    "#     A += pMatTri\n",
    "#     A += hMatTri\n",
    "    \n",
    "#     ABinary = np.where(A > 0, 1, 0)\n",
    "    \n",
    "#     plt.matshow(ABinary, cmap='binary')\n",
    "#     plt.title(f'Raw scPore-C Contacts {pKey}')\n",
    "#     plt.text((n//2), n+100, f\"Raw scHi-C Contacts {hKey}\", \n",
    "#              horizontalalignment='center', fontsize=12)\n",
    "    \n",
    "    \n",
    "#     plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# psuedoPoreC = np.zeros(poreC['run01'].shape)\n",
    "\n",
    "# for k,v in poreC.items():\n",
    "#     psuedoPoreC += v.matrix(balance=True)[:]\n",
    "    \n",
    "\n",
    "# # kinda hacky\n",
    "# psuedoHiC = np.zeros(hiC[list(hiC.keys())[0]].shape)\n",
    "    \n",
    "# for k,v in hiC.items():\n",
    "#     psuedoHiC += v.matrix(balance=True)[:]\n",
    "\n",
    "# print(f\"{psuedoPoreC.shape=}\")\n",
    "# print(f\"{psuedoHiC.shape=}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pMatBinary = np.where(psuedoPoreC > 0, 1, 0)\n",
    "# hMatBinary = np.where(psuedoHiC > 0, 1, 0)\n",
    "\n",
    "# A = np.zeros(psuedoHiC.shape)\n",
    "\n",
    "# # truncate the y chrome of the porec data\n",
    "# n = psuedoHiC.shape[0]\n",
    "# pMatBinary = pMatBinary[0:n, 0:n]\n",
    "\n",
    "# # get upper and lower triangles\n",
    "# pMatTri = np.triu(pMatBinary)\n",
    "# hMatTri = np.tril(hMatBinary)\n",
    "\n",
    "# A += pMatTri\n",
    "# A += hMatTri\n",
    "\n",
    "# ABinary = np.where(A > 0, 1, 0)\n",
    "\n",
    "# plt.matshow(ABinary, cmap='binary')\n",
    "# plt.title(f'Raw scPore-C Contacts Ensemble')\n",
    "# plt.text((n//2), n+100, f\"Raw scHi-C Contacts Ensemble\", \n",
    "#          horizontalalignment='center', fontsize=12)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plt.close()\n",
    "# plt.rcParams['figure.dpi'] = 200\n",
    "# plt.rcParams['figure.figsize'] = 3, 3\n",
    "\n",
    "# # chromosome 2\n",
    "# zippedKeys = list(zip(poreC.keys(), hiC.keys()))\n",
    "\n",
    "# for pKey, hKey in zippedKeys:\n",
    "    \n",
    "#     print(\"------------------------\")\n",
    "#     print(f\"{pKey=}\")\n",
    "#     print(f\"{hKey=}\")\n",
    "#     pMat = poreC[pKey].matrix(balance=True).fetch('NC_000068.8')[:]\n",
    "#     hMat = hiC[hKey].matrix(balance=True).fetch('chr2')[:]\n",
    "    \n",
    "#     print(f\"{pMat.shape}\")\n",
    "#     print(f\"{hMat.shape}\")\n",
    "    \n",
    "    \n",
    "#     # binarize both\n",
    "#     pMatBinary = np.where(pMat > 0, 1, 0)\n",
    "#     hMatBinary = np.where(hMat > 0, 1, 0)\n",
    "    \n",
    "#     # get upper and lower triangles\n",
    "#     pMatTri = np.triu(pMatBinary)\n",
    "#     hMatTri = np.tril(hMatBinary)\n",
    "    \n",
    "#     A = np.zeros(pMat.shape)\n",
    "    \n",
    "#     A += pMatTri\n",
    "#     A += hMatTri\n",
    "    \n",
    "#     ABinary = np.where(A > 0, 1, 0)\n",
    "    \n",
    "#     plt.matshow(ABinary, cmap='binary')\n",
    "#     plt.title(f'Raw scPore-C Contacts {pKey}')\n",
    "#     plt.text((n//2), n+100, f\"Raw scHi-C Contacts {hKey}\", \n",
    "#              horizontalalignment='center', fontsize=12)\n",
    "    \n",
    "#     plt.show()\n",
    "    \n",
    "#     break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# A = np.zeros(psuedoHiC.shape)\n",
    "\n",
    "# # truncate the y chrome of the porec data\n",
    "# n = psuedoHiC.shape[0]\n",
    "# psuedoPoreC = psuedoPoreC[0:n, 0:n]\n",
    "\n",
    "# # get upper and lower triangles\n",
    "# pMatTri = np.triu(psuedoPoreC)\n",
    "# hMatTri = np.tril(psuedoHiC)\n",
    "\n",
    "# A += pMatTri\n",
    "# A += hMatTri\n",
    "\n",
    "# plt.matshow(A, cmap='Reds')\n",
    "# plt.title(f'Raw scPore-C Contacts Ensemble')\n",
    "# plt.text((n//2), n+100, f\"Raw scHi-C Contacts Ensemble\", \n",
    "#          horizontalalignment='center', fontsize=12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
