{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import cooler\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors\n",
    "from matplotlib.patches import Patch\n",
    "from matplotlib.offsetbox import AnchoredText\n",
    "import scipy.io\n",
    "import gget\n",
    "\n",
    "import os\n",
    "import sys\n",
    "from importlib import reload\n",
    "\n",
    "import filters\n",
    "import pore_c_utils as pcu\n",
    "import binning_utils as binning\n",
    "import plotting as hicPlot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4DNFICXJQ3PA.tsv\n",
      "4DNFIW5IZKYG.bw\n",
      "4DNEX1QL8KF2_processed_files_2022-05-30-10h-02m.tsv\n",
      "4DNFIC269AEU.tsv\n",
      "4DNFI2HFSUA8.bam\n"
     ]
    }
   ],
   "source": [
    "for f in os.listdir(dirpath):\n",
    "    print(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df.shape=(168207, 19)\n",
      "Index(['transcript_id', 'gene_id', 'length', 'effective_length',\n",
      "       'expected_count', 'TPM', 'FPKM', 'IsoPct', 'posterior_mean_count',\n",
      "       'posterior_standard_deviation_of_count', 'pme_TPM', 'pme_FPKM',\n",
      "       'IsoPct_from_pme_TPM', 'TPM_ci_lower_bound', 'TPM_ci_upper_bound',\n",
      "       'TPM_coefficient_of_quartile_variation', 'FPKM_ci_lower_bound',\n",
      "       'FPKM_ci_upper_bound', 'FPKM_coefficient_of_quartile_variation'],\n",
      "      dtype='object')\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>length</th>\n",
       "      <th>effective_length</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>IsoPct</th>\n",
       "      <th>posterior_mean_count</th>\n",
       "      <th>posterior_standard_deviation_of_count</th>\n",
       "      <th>pme_TPM</th>\n",
       "      <th>pme_FPKM</th>\n",
       "      <th>IsoPct_from_pme_TPM</th>\n",
       "      <th>TPM_ci_lower_bound</th>\n",
       "      <th>TPM_ci_upper_bound</th>\n",
       "      <th>TPM_coefficient_of_quartile_variation</th>\n",
       "      <th>FPKM_ci_lower_bound</th>\n",
       "      <th>FPKM_ci_upper_bound</th>\n",
       "      <th>FPKM_coefficient_of_quartile_variation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10000</td>\n",
       "      <td>10000</td>\n",
       "      <td>72</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10001</td>\n",
       "      <td>10001</td>\n",
       "      <td>73</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10002</td>\n",
       "      <td>10002</td>\n",
       "      <td>73</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10003</td>\n",
       "      <td>10003</td>\n",
       "      <td>75</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10004</td>\n",
       "      <td>10004</td>\n",
       "      <td>78</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  transcript_id gene_id  length  effective_length  expected_count  TPM  FPKM  \\\n",
       "0         10000   10000      72               0.0             0.0  0.0   0.0   \n",
       "1         10001   10001      73               0.0             0.0  0.0   0.0   \n",
       "2         10002   10002      73               0.0             0.0  0.0   0.0   \n",
       "3         10003   10003      75               0.0             0.0  0.0   0.0   \n",
       "4         10004   10004      78               0.0             0.0  0.0   0.0   \n",
       "\n",
       "   IsoPct  posterior_mean_count  posterior_standard_deviation_of_count  \\\n",
       "0     0.0                   0.0                                    0.0   \n",
       "1     0.0                   0.0                                    0.0   \n",
       "2     0.0                   0.0                                    0.0   \n",
       "3     0.0                   0.0                                    0.0   \n",
       "4     0.0                   0.0                                    0.0   \n",
       "\n",
       "   pme_TPM  pme_FPKM  IsoPct_from_pme_TPM  TPM_ci_lower_bound  \\\n",
       "0      0.0       0.0                  0.0                 0.0   \n",
       "1      0.0       0.0                  0.0                 0.0   \n",
       "2      0.0       0.0                  0.0                 0.0   \n",
       "3      0.0       0.0                  0.0                 0.0   \n",
       "4      0.0       0.0                  0.0                 0.0   \n",
       "\n",
       "   TPM_ci_upper_bound  TPM_coefficient_of_quartile_variation  \\\n",
       "0                 0.0                                    0.0   \n",
       "1                 0.0                                    0.0   \n",
       "2                 0.0                                    0.0   \n",
       "3                 0.0                                    0.0   \n",
       "4                 0.0                                    0.0   \n",
       "\n",
       "   FPKM_ci_lower_bound  FPKM_ci_upper_bound  \\\n",
       "0                  0.0                  0.0   \n",
       "1                  0.0                  0.0   \n",
       "2                  0.0                  0.0   \n",
       "3                  0.0                  0.0   \n",
       "4                  0.0                  0.0   \n",
       "\n",
       "   FPKM_coefficient_of_quartile_variation  \n",
       "0                                     0.0  \n",
       "1                                     0.0  \n",
       "2                                     0.0  \n",
       "3                                     0.0  \n",
       "4                                     0.0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dirpath = \"/nfs/turbo/umms-indikar/shared/projects/poreC/kauffmanRNA/\"\n",
    "file = \"4DNFICXJQ3PA.tsv\"\n",
    "\n",
    "fullpath = f\"{dirpath}{file}\"\n",
    "    \n",
    "df = pd.read_csv(fullpath, sep='\\t')\n",
    "print(f\"{df.shape=}\")\n",
    "print(df.columns)\n",
    "print()\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df.shape=(168207, 19)\n",
      "Index(['transcript_id', 'gene_id', 'length', 'effective_length',\n",
      "       'expected_count', 'TPM', 'FPKM', 'IsoPct', 'posterior_mean_count',\n",
      "       'posterior_standard_deviation_of_count', 'pme_TPM', 'pme_FPKM',\n",
      "       'IsoPct_from_pme_TPM', 'TPM_ci_lower_bound', 'TPM_ci_upper_bound',\n",
      "       'TPM_coefficient_of_quartile_variation', 'FPKM_ci_lower_bound',\n",
      "       'FPKM_ci_upper_bound', 'FPKM_coefficient_of_quartile_variation'],\n",
      "      dtype='object')\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>length</th>\n",
       "      <th>effective_length</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>IsoPct</th>\n",
       "      <th>posterior_mean_count</th>\n",
       "      <th>posterior_standard_deviation_of_count</th>\n",
       "      <th>pme_TPM</th>\n",
       "      <th>pme_FPKM</th>\n",
       "      <th>IsoPct_from_pme_TPM</th>\n",
       "      <th>TPM_ci_lower_bound</th>\n",
       "      <th>TPM_ci_upper_bound</th>\n",
       "      <th>TPM_coefficient_of_quartile_variation</th>\n",
       "      <th>FPKM_ci_lower_bound</th>\n",
       "      <th>FPKM_ci_upper_bound</th>\n",
       "      <th>FPKM_coefficient_of_quartile_variation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10000</td>\n",
       "      <td>10000</td>\n",
       "      <td>72</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10001</td>\n",
       "      <td>10001</td>\n",
       "      <td>73</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10002</td>\n",
       "      <td>10002</td>\n",
       "      <td>73</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10003</td>\n",
       "      <td>10003</td>\n",
       "      <td>75</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10004</td>\n",
       "      <td>10004</td>\n",
       "      <td>78</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  transcript_id gene_id  length  effective_length  expected_count  TPM  FPKM  \\\n",
       "0         10000   10000      72               0.0             0.0  0.0   0.0   \n",
       "1         10001   10001      73               0.0             0.0  0.0   0.0   \n",
       "2         10002   10002      73               0.0             0.0  0.0   0.0   \n",
       "3         10003   10003      75               0.0             0.0  0.0   0.0   \n",
       "4         10004   10004      78               0.0             0.0  0.0   0.0   \n",
       "\n",
       "   IsoPct  posterior_mean_count  posterior_standard_deviation_of_count  \\\n",
       "0     0.0                   0.0                                    0.0   \n",
       "1     0.0                   0.0                                    0.0   \n",
       "2     0.0                   0.0                                    0.0   \n",
       "3     0.0                   0.0                                    0.0   \n",
       "4     0.0                   0.0                                    0.0   \n",
       "\n",
       "   pme_TPM  pme_FPKM  IsoPct_from_pme_TPM  TPM_ci_lower_bound  \\\n",
       "0      0.0       0.0                  0.0                 0.0   \n",
       "1      0.0       0.0                  0.0                 0.0   \n",
       "2      0.0       0.0                  0.0                 0.0   \n",
       "3      0.0       0.0                  0.0                 0.0   \n",
       "4      0.0       0.0                  0.0                 0.0   \n",
       "\n",
       "   TPM_ci_upper_bound  TPM_coefficient_of_quartile_variation  \\\n",
       "0                 0.0                                    0.0   \n",
       "1                 0.0                                    0.0   \n",
       "2                 0.0                                    0.0   \n",
       "3                 0.0                                    0.0   \n",
       "4                 0.0                                    0.0   \n",
       "\n",
       "   FPKM_ci_lower_bound  FPKM_ci_upper_bound  \\\n",
       "0                  0.0                  0.0   \n",
       "1                  0.0                  0.0   \n",
       "2                  0.0                  0.0   \n",
       "3                  0.0                  0.0   \n",
       "4                  0.0                  0.0   \n",
       "\n",
       "   FPKM_coefficient_of_quartile_variation  \n",
       "0                                     0.0  \n",
       "1                                     0.0  \n",
       "2                                     0.0  \n",
       "3                                     0.0  \n",
       "4                                     0.0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dirpath = \"/nfs/turbo/umms-indikar/shared/projects/poreC/nagano_2017_rnaseq/\"\n",
    "file = \"4DNFI8CSCJWM.tsv\"\n",
    "\n",
    "fullpath = f\"{dirpath}{file}\"\n",
    "    \n",
    "df = pd.read_csv(fullpath, sep='\\t')\n",
    "print(f\"{df.shape=}\")\n",
    "print(df.columns)\n",
    "print()\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ENSMUSG00000053477.17    62\n",
       "ENSMUSG00000099722.3     58\n",
       "ENSMUSG00000074415.14    48\n",
       "ENSMUSG00000069601.14    47\n",
       "ENSMUSG00000024921.17    46\n",
       "                         ..\n",
       "61448                     1\n",
       "61449                     1\n",
       "6145                      1\n",
       "61450                     1\n",
       "gSpikein_phiX174          1\n",
       "Name: gene_id, Length: 81881, dtype: int64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['gene_id'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mon May 30 15:03:09 2022 INFO Returning only reviewed UniProt results for Ensembl ID ENSMUSG00000053477.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>ncbi_gene_id</th>\n",
       "      <th>species</th>\n",
       "      <th>assembly_name</th>\n",
       "      <th>primary_gene_name</th>\n",
       "      <th>ensembl_gene_name</th>\n",
       "      <th>synonyms</th>\n",
       "      <th>parent_gene</th>\n",
       "      <th>protein_names</th>\n",
       "      <th>ensembl_description</th>\n",
       "      <th>uniprot_description</th>\n",
       "      <th>ncbi_description</th>\n",
       "      <th>object_type</th>\n",
       "      <th>biotype</th>\n",
       "      <th>canonical_transcript</th>\n",
       "      <th>seq_region_name</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000053477</th>\n",
       "      <td>ENSMUSG00000053477</td>\n",
       "      <td>Q60722</td>\n",
       "      <td>21413</td>\n",
       "      <td>mus_musculus</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>Tcf4</td>\n",
       "      <td>Tcf4</td>\n",
       "      <td>[5730422P05Rik, ASP-I2, E2-2, E2.2, ITF-2, ITF...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Transcription factor 4 (TCF-4) (Class A helix-...</td>\n",
       "      <td>transcription factor 4 [Source:MGI Symbol;Acc:...</td>\n",
       "      <td>FUNCTION: Transcription factor that binds to t...</td>\n",
       "      <td>Enables several functions, including DNA-bindi...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSMUST00000066717.12</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>69476427</td>\n",
       "      <td>69822150</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            ensembl_id uniprot_id ncbi_gene_id       species  \\\n",
       "ENSMUSG00000053477  ENSMUSG00000053477     Q60722        21413  mus_musculus   \n",
       "\n",
       "                   assembly_name primary_gene_name ensembl_gene_name  \\\n",
       "ENSMUSG00000053477        GRCm39              Tcf4              Tcf4   \n",
       "\n",
       "                                                             synonyms  \\\n",
       "ENSMUSG00000053477  [5730422P05Rik, ASP-I2, E2-2, E2.2, ITF-2, ITF...   \n",
       "\n",
       "                   parent_gene  \\\n",
       "ENSMUSG00000053477         NaN   \n",
       "\n",
       "                                                        protein_names  \\\n",
       "ENSMUSG00000053477  Transcription factor 4 (TCF-4) (Class A helix-...   \n",
       "\n",
       "                                                  ensembl_description  \\\n",
       "ENSMUSG00000053477  transcription factor 4 [Source:MGI Symbol;Acc:...   \n",
       "\n",
       "                                                  uniprot_description  \\\n",
       "ENSMUSG00000053477  FUNCTION: Transcription factor that binds to t...   \n",
       "\n",
       "                                                     ncbi_description  \\\n",
       "ENSMUSG00000053477  Enables several functions, including DNA-bindi...   \n",
       "\n",
       "                   object_type         biotype   canonical_transcript  \\\n",
       "ENSMUSG00000053477        Gene  protein_coding  ENSMUST00000066717.12   \n",
       "\n",
       "                   seq_region_name strand     start       end  \n",
       "ENSMUSG00000053477              18      1  69476427  69822150  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gget.info(\"ENSMUSG00000053477.17\", expand)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mon May 30 15:01:54 2022 ERROR Ensembl ID '10000' not found. Please double-check spelling/arguments and try again.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['__bool__', '__class__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__']\n"
     ]
    }
   ],
   "source": [
    "for idx, row in df.iterrows():\n",
    "    \n",
    "    gId = row['gene_id']\n",
    "    \n",
    "    t = gget.info(gId, expand=True)\n",
    "    print(dir(t))\n",
    "    \n",
    "    break\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
