{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import cooler\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors\n",
    "from matplotlib.patches import Patch\n",
    "import scipy.io\n",
    "import pysam\n",
    "\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "from scipy import stats\n",
    "import networkx as nx\n",
    "from sklearn.metrics import jaccard_score\n",
    "\n",
    "import os\n",
    "import sys\n",
    "from importlib import reload\n",
    "\n",
    "import filters\n",
    "import pore_c_utils as pcu\n",
    "import binning_utils as binning\n",
    "import plotting as hicPlot\n",
    "import nanoporeDataModel as ndm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/sw/arcts/centos7/python3.8-anaconda/2020.07/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3343: DtypeWarning: Columns (39) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df.shape=(636130, 65)\n",
      "Index(['index', 'read_name', 'read_length', 'read_idx', 'contact_is_direct',\n",
      "       'contact_is_cis', 'contact_read_distance', 'contact_genome_distance',\n",
      "       'contact_fragment_adjacent', 'contact_fragment_distance',\n",
      "       'haplotype_pair_type', 'align1_align_idx', 'align1_chrom',\n",
      "       'align1_start', 'align1_end', 'align1_strand', 'align1_mapping_quality',\n",
      "       'align1_align_score', 'align1_align_base_qscore', 'align1_phase_set',\n",
      "       'align1_phase_qual', 'align1_haplotype', 'align1_fragment_id',\n",
      "       'align1_fragment_start', 'align1_fragment_end', 'align2_align_idx',\n",
      "       'align2_chrom', 'align2_start', 'align2_end', 'align2_strand',\n",
      "       'align2_mapping_quality', 'align2_align_score',\n",
      "       'align2_align_base_qscore', 'align2_phase_set', 'align2_haplotype',\n",
      "       'align2_fragment_id', 'align2_fragment_start', 'align2_fragment_end',\n",
      "       'batch_id', 'align1_chromosome_name', 'align1_chrom_length',\n",
      "       'align1_genbank', 'align1_refSeq', 'align1_chrom_end',\n",
      "       'align1_chrom_start', 'align2_chromosome_name', 'align2_chrom_length',\n",
      "       'align2_genbank', 'align2_refSeq', 'align2_chrom_end',\n",
      "       'align2_chrom_start', 'align1_absolute_start', 'align2_absolute_start',\n",
      "       'align1_absolute_end', 'align2_absolute_end',\n",
      "       'align1_absolute_midpoint', 'align2_absolute_midpoint',\n",
      "       'mean_mapping_quality', 'contact_count', 'align1_fragcount',\n",
      "       'align2_fragcount', 'align1_n_reads', 'align2_n_reads',\n",
      "       'contact_has_support', 'cell'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>read_name</th>\n",
       "      <th>read_length</th>\n",
       "      <th>read_idx</th>\n",
       "      <th>contact_is_direct</th>\n",
       "      <th>contact_is_cis</th>\n",
       "      <th>contact_read_distance</th>\n",
       "      <th>contact_genome_distance</th>\n",
       "      <th>contact_fragment_adjacent</th>\n",
       "      <th>contact_fragment_distance</th>\n",
       "      <th>...</th>\n",
       "      <th>align1_absolute_midpoint</th>\n",
       "      <th>align2_absolute_midpoint</th>\n",
       "      <th>mean_mapping_quality</th>\n",
       "      <th>contact_count</th>\n",
       "      <th>align1_fragcount</th>\n",
       "      <th>align2_fragcount</th>\n",
       "      <th>align1_n_reads</th>\n",
       "      <th>align2_n_reads</th>\n",
       "      <th>contact_has_support</th>\n",
       "      <th>cell</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>27771</td>\n",
       "      <td>132f1be4-df3c-4a1f-bc53-40df856b0c42</td>\n",
       "      <td>4312</td>\n",
       "      <td>35703</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1619</td>\n",
       "      <td>11803</td>\n",
       "      <td>False</td>\n",
       "      <td>12375</td>\n",
       "      <td>...</td>\n",
       "      <td>2630943949</td>\n",
       "      <td>2630956324</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>run01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>206517</td>\n",
       "      <td>b0e13742-1d17-47d9-a017-668f37b6788f</td>\n",
       "      <td>2616</td>\n",
       "      <td>7786</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1432</td>\n",
       "      <td>714701</td>\n",
       "      <td>False</td>\n",
       "      <td>715354</td>\n",
       "      <td>...</td>\n",
       "      <td>2630229012</td>\n",
       "      <td>2630944366</td>\n",
       "      <td>96.0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>run01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>117363</td>\n",
       "      <td>330b862a-138c-48d7-9534-be879100a320</td>\n",
       "      <td>1022</td>\n",
       "      <td>31146</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>1378</td>\n",
       "      <td>False</td>\n",
       "      <td>1844</td>\n",
       "      <td>...</td>\n",
       "      <td>2630227938</td>\n",
       "      <td>2630229781</td>\n",
       "      <td>49.5</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>run01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>468866</td>\n",
       "      <td>af44682d-de48-467d-8cd2-f79d45b85f69</td>\n",
       "      <td>1364</td>\n",
       "      <td>36416</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>254</td>\n",
       "      <td>729058</td>\n",
       "      <td>False</td>\n",
       "      <td>729603</td>\n",
       "      <td>...</td>\n",
       "      <td>2630227390</td>\n",
       "      <td>2630956993</td>\n",
       "      <td>47.5</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>run01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>180875</td>\n",
       "      <td>7470a373-22cb-49b2-a0df-9e5279d9559c</td>\n",
       "      <td>3386</td>\n",
       "      <td>7678</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>866</td>\n",
       "      <td>716159</td>\n",
       "      <td>False</td>\n",
       "      <td>716559</td>\n",
       "      <td>...</td>\n",
       "      <td>2630227390</td>\n",
       "      <td>2630943949</td>\n",
       "      <td>50.5</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>run01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    index                             read_name  read_length  read_idx  \\\n",
       "0   27771  132f1be4-df3c-4a1f-bc53-40df856b0c42         4312     35703   \n",
       "1  206517  b0e13742-1d17-47d9-a017-668f37b6788f         2616      7786   \n",
       "2  117363  330b862a-138c-48d7-9534-be879100a320         1022     31146   \n",
       "3  468866  af44682d-de48-467d-8cd2-f79d45b85f69         1364     36416   \n",
       "4  180875  7470a373-22cb-49b2-a0df-9e5279d9559c         3386      7678   \n",
       "\n",
       "   contact_is_direct  contact_is_cis  contact_read_distance  \\\n",
       "0              False            True                   1619   \n",
       "1              False            True                   1432   \n",
       "2               True            True                      1   \n",
       "3               True            True                    254   \n",
       "4              False            True                    866   \n",
       "\n",
       "   contact_genome_distance  contact_fragment_adjacent  \\\n",
       "0                    11803                      False   \n",
       "1                   714701                      False   \n",
       "2                     1378                      False   \n",
       "3                   729058                      False   \n",
       "4                   716159                      False   \n",
       "\n",
       "   contact_fragment_distance  ... align1_absolute_midpoint  \\\n",
       "0                      12375  ...               2630943949   \n",
       "1                     715354  ...               2630229012   \n",
       "2                       1844  ...               2630227938   \n",
       "3                     729603  ...               2630227390   \n",
       "4                     716559  ...               2630227390   \n",
       "\n",
       "   align2_absolute_midpoint mean_mapping_quality  contact_count  \\\n",
       "0                2630956324                 50.0              1   \n",
       "1                2630944366                 96.0              1   \n",
       "2                2630229781                 49.5              1   \n",
       "3                2630956993                 47.5              1   \n",
       "4                2630943949                 50.5              1   \n",
       "\n",
       "   align1_fragcount  align2_fragcount  align1_n_reads  align2_n_reads  \\\n",
       "0                 4                 3             NaN             NaN   \n",
       "1                 4                 2             NaN             NaN   \n",
       "2                 3                 2             NaN             NaN   \n",
       "3                 2                 3             NaN             NaN   \n",
       "4                 2                 4             NaN             NaN   \n",
       "\n",
       "   contact_has_support   cell  \n",
       "0                    1  run01  \n",
       "1                    1  run01  \n",
       "2                    1  run01  \n",
       "3                    1  run01  \n",
       "4                    1  run01  \n",
       "\n",
       "[5 rows x 65 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filepath = \"filtered_data/clique1MBfiltered.csv\"\n",
    "\n",
    "df = pd.read_csv(filepath)\n",
    "print(f\"{df.shape=}\")\n",
    "print(df.columns)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "runId='run01'\n",
      "cellDf.shape=(6226401, 9)\n",
      "cellDf.shape=(6226401, 9)\n",
      "runId='run02'\n",
      "cellDf.shape=(5946864, 9)\n",
      "cellDf.shape=(5946864, 9)\n",
      "runId='run04'\n",
      "cellDf.shape=(8193791, 9)\n",
      "cellDf.shape=(8193791, 9)\n",
      "runId='run05'\n",
      "cellDf.shape=(2802500, 9)\n",
      "cellDf.shape=(2802500, 9)\n",
      "runId='run06'\n",
      "cellDf.shape=(3650575, 9)\n",
      "cellDf.shape=(3650575, 9)\n",
      "runId='run07'\n",
      "cellDf.shape=(4258406, 9)\n",
      "cellDf.shape=(4258406, 9)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "dict_keys(['run01', 'run02', 'run04', 'run05', 'run06', 'run07'])"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bampathDir = '/nfs/turbo/umms-indikar/shared/tools/Pore-C-Snakemake/results_ES5to12/mapping/'\n",
    "\n",
    "def sam2SequenceFrame(bam):\n",
    "    \"\"\" Convert a bam file to a table \"\"\"\n",
    "    newRows = []\n",
    "    for align in bam:\n",
    "        read_name, read_idx, align_idx = align.query_name.split(\":\")\n",
    "        row = {\n",
    "            'read_name' : read_name,\n",
    "            'read_idx' : read_idx,\n",
    "            'align_idx' : align_idx, \n",
    "            'seq' : align.seq,\n",
    "            'read_length' : align.infer_read_length(),\n",
    "            'read_start' : align.query_alignment_start,\n",
    "            'read_end' : align.query_alignment_end,\n",
    "        }\n",
    "        newRows.append(row)\n",
    "    return pd.DataFrame(newRows)\n",
    "    \n",
    "\n",
    "seqs = {}\n",
    "\n",
    "for runId in df['cell'].unique():\n",
    "    \n",
    "    print(f\"{runId=}\")\n",
    "    dfList = []\n",
    "    \n",
    "    for file in os.listdir(bampathDir):\n",
    "        if runId in file and file.endswith('.coord_sort.bam'):\n",
    "            filepath = f\"{bampathDir}{file}\"        \n",
    "\n",
    "            runId = file.split(\"_\")[1]\n",
    "\n",
    "            bam = pysam.AlignmentFile(filepath)\n",
    "            readDf = sam2SequenceFrame(bam)\n",
    "\n",
    "            readDf['file'] = file\n",
    "            readDf['runId'] = runId\n",
    "            \n",
    "            dfList.append(readDf)\n",
    "    cellDf = pd.concat(dfList, ignore_index=True)\n",
    "    print(f\"{cellDf.shape=}\")\n",
    "    seqs[runId] = cellDf\n",
    "    \n",
    "    \n",
    "seqs.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "runId='run01'\n",
      "data.shape=(6226401, 9)\n",
      "data.shape=(322752, 9)\n",
      "\n",
      "runId='run02'\n",
      "data.shape=(5946864, 9)\n",
      "data.shape=(170051, 9)\n",
      "\n",
      "runId='run04'\n",
      "data.shape=(8193791, 9)\n",
      "data.shape=(214386, 9)\n",
      "\n",
      "runId='run05'\n",
      "data.shape=(2802500, 9)\n",
      "data.shape=(233556, 9)\n",
      "\n",
      "runId='run06'\n",
      "data.shape=(3650575, 9)\n",
      "data.shape=(237408, 9)\n",
      "\n",
      "runId='run07'\n",
      "data.shape=(4258406, 9)\n",
      "data.shape=(408510, 9)\n",
      "\n",
      "done\n"
     ]
    }
   ],
   "source": [
    "for runId, data in seqs.items():\n",
    "    print(f\"{runId=}\")\n",
    "    print(f\"{data.shape=}\")\n",
    "    data = data[data['read_name'].isin(df['read_name'])]\n",
    "    print(f\"{data.shape=}\")\n",
    "    print()\n",
    "    \n",
    "    seqs[runId] = data\n",
    "    \n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saved: filtered_data/run01_sequences.csv\n",
      "saved: filtered_data/run02_sequences.csv\n",
      "saved: filtered_data/run04_sequences.csv\n",
      "saved: filtered_data/run05_sequences.csv\n",
      "saved: filtered_data/run06_sequences.csv\n",
      "saved: filtered_data/run07_sequences.csv\n",
      "done.\n"
     ]
    }
   ],
   "source": [
    "for runId, data in seqs.items():\n",
    "    \n",
    "    filepath = f'filtered_data/{runId}_sequences.csv'\n",
    "    \n",
    "    data.to_csv(filepath, index=False)\n",
    "    print(f'saved: {filepath}')\n",
    "    \n",
    "print('done.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
